{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Study session 11 - SQL\n",
    "\n",
    "#### BIOINF 575 - Fall 2023\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# put the import statements at the top of the notebook\n",
    "\n",
    "# bring in the functionality from the sqlite3 module\n",
    "# specifically the connect function\n",
    "\n",
    "from sqlite3 import connect\n",
    "\n",
    "# import pandas so you can see the select results as a dataframe\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "#### Create a database for gene annotations (gene ontology - GO -  terms annotations)\n",
    "\n",
    "Create a SQLite database with a genes table, a GO_terms table, and a gene_GO_term table.   \n",
    "The genes table should have the gene id, gene symbol and the gene description.     \n",
    "The GO_terms table should have the GO term id and the GO term name/description.     \n",
    "The gene to GO term table should have the gene id and the GO term id.     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "# Write the create table statements here - all strings\n",
    "\n",
    "#-------\n",
    "create_genes = '''\n",
    "CREATE TABLE IF NOT EXISTS genes (\n",
    "      gene_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "      gene_symbol TEXT NOT NULL,\n",
    "      gene_description TEXT NOT NULL\n",
    "    );\n",
    "'''\n",
    "#-------\n",
    "\n",
    "create_GO_terms = '''\n",
    "CREATE TABLE IF NOT EXISTS GO_terms (\n",
    "      GO_term_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "      name TEXT NOT NULL\n",
    "    );\n",
    "'''\n",
    "#-------\n",
    "\n",
    "create_gene_GO_term = '''\n",
    "CREATE TABLE IF NOT EXISTS gene_GO_term (\n",
    "      gene_id INTEGER NOT NULL,                             -- REFERENCES  gene_id in the genes table\n",
    "      GO_term_id INTEGER NOT NULL,                          -- REFERENCES  GO_term_id in the GO_terms table\n",
    "      \n",
    "      PRIMARY KEY (gene_id, GO_term_id),                    -- we set up a multi-column primary key - table constraint\n",
    "                                                                -- the two columns uniquely indetify a row \n",
    "                                                                -- this will automatically create an index in the database\n",
    "      FOREIGN KEY (gene_id) REFERENCES  genes  (gene_id),\n",
    "      FOREIGN KEY (GO_term_id) REFERENCES  GO_terms  (GO_term_id)\n",
    "    );\n",
    "'''\n",
    "#-------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we do the same thing multiple times - we can use a function\n",
    "\n",
    "def run_create_table(sql, cursor, connection):\n",
    "    try:\n",
    "        cursor.execute(sql)\n",
    "    except connection.DatabaseError:\n",
    "        print(\"Creating the table resulted in a database error!\")\n",
    "        connection.rollback()\n",
    "        raise\n",
    "    else:\n",
    "        connection.commit()\n",
    "    finally:\n",
    "        print(\"done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect to a database ... if the file does not exist it will be created\n",
    "connection = connect(\"gene_annotations.sqlite\")\n",
    "cursor = connection.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sqlite_master table - should be empty at this point\n",
    "\n",
    "select_master = \"SELECT name, type FROM sqlite_master;\"\n",
    "cursor.execute(select_master)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n",
      "done!\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "# run the commands\n",
    "run_create_table(create_genes, cursor, connection)\n",
    "run_create_table(create_GO_terms, cursor, connection)\n",
    "run_create_table(create_gene_GO_term, cursor, connection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('genes', 'table'),\n",
       " ('sqlite_sequence', 'table'),\n",
       " ('GO_terms', 'table'),\n",
       " ('gene_GO_term', 'table'),\n",
       " ('sqlite_autoindex_gene_GO_term_1', 'index')]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sqlite_master table\n",
    "\n",
    "# select_master = \"SELECT name, type FROM sqlite_master;\"\n",
    "cursor.execute(select_master)\n",
    "cursor.fetchall()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "#### Populate a database \n",
    "\n",
    "Insert data into tables in the GO annotations database you just created.   \n",
    "Use the data from the file `genes_info1.txt`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "# Write the insert statements here - all strings\n",
    "# because we insert values in all columns we do not need to specify the columns\n",
    "\n",
    "insert_gene = \"INSERT INTO genes VALUES(?,?,?);\"\n",
    "insert_GO_term = \"INSERT INTO GO_terms VALUES(?,?);\"\n",
    "insert_gene_GO_term = \"INSERT INTO gene_GO_term VALUES(?,?);\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"genes_info1.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "GO_terms_dict = {}\n",
    "GO_term_index = 0\n",
    "\n",
    "with open(file_name) as genes_info:\n",
    "    # read the header line and ignore it\n",
    "    header_line = genes_info.readline() \n",
    "    \n",
    "    # read the rest of the lines and process them\n",
    "    for line in genes_info: \n",
    "        # break the line by tab and take the information needed in the table\n",
    "        line_list = line.strip().split(\"\\t\")\n",
    "        \n",
    "        # we put the gene id in a variable because we need it in gene_GO_term table\n",
    "        gene_id = line_list[0]\n",
    "        \n",
    "        # we take the gene id, symbol, description, \n",
    "        # which are the first 3 elements in the list - see header_line to confirm\n",
    "        # this will be a row in the genes table\n",
    "        gene_row = line_list[:3]\n",
    "        cursor.execute(insert_gene, gene_row)\n",
    "        \n",
    "        # we take the go terms which are last in the list\n",
    "        # we breake them up into a list - they are separated by \";\"\n",
    "        GO_terms_list = line_list[-1].split(\";\")\n",
    "        \n",
    "        # we go through the list of go terms to create rows for the GO_terms table\n",
    "        # we need to keep track of the GO terms so we do not add them multiple times in the table\n",
    "        # one GO term can be associated with multiple genes and one genes can be associated with multiple GO terms\n",
    "        # a dictionary has unique keys and can store values foe each key\n",
    "        # we will use that to keep track of our GO terms\n",
    "        for GO_term in GO_terms_list:\n",
    "            # remove whitespace from the beginning and end of the GO_term\n",
    "            GO_term_name = GO_term.strip()\n",
    "            \n",
    "            # if we did not see this GO term before we add it to the dictionary and create an ID for it\n",
    "            # otherwise we get the id of the term from the dictionary\n",
    "            if GO_term_name not in GO_terms_dict:\n",
    "                GO_term_index += 1\n",
    "                GO_term_id = GO_term_index\n",
    "                GO_terms_dict[GO_term_name] = GO_term_index\n",
    "                # we will add this new GO term data to the GO_term table so prepare a row\n",
    "                GO_term_row = (GO_term_id, GO_term_name)\n",
    "                cursor.execute(insert_GO_term, GO_term_row)\n",
    "\n",
    "            else:\n",
    "                GO_term_id = GO_terms_dict[GO_term_name]\n",
    "            \n",
    "            # we add the connection between the gene and the GO term to the gene_GO_term table\n",
    "            # we prepare a row\n",
    "            gene_GO_row = (gene_id, GO_term_id)\n",
    "            cursor.execute(insert_gene_GO_term, gene_GO_row)\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we check the tables and if all is well, we can commit these changes\n",
    "\n",
    "select_genes = \"SELECT * FROM genes;\"\n",
    "select_GO_terms = \"SELECT * FROM GO_terms;\"\n",
    "select_gene_GO = \"SELECT * FROM gene_GO_term;\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 'BRCA1', 'BRCA1 DNA repair associated'),\n",
       " (675, 'BRCA2', 'BRCA2 DNA repair associated'),\n",
       " (7040, 'TGFB1', 'transforming growth factor beta 1'),\n",
       " (7157, 'TP53', 'tumor protein p53'),\n",
       " (10296, 'MAEA', 'macrophage erythroblast attacher, E3 ubiquitin ligase'),\n",
       " (64926, 'RASAL3', 'RAS protein activator like 3')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_genes)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1, 'intrinsic apoptotic signaling pathway in response to DNA damage'),\n",
       " (2, 'transcription cis-regulatory region binding'),\n",
       " (3, 'transcription coactivator activity'),\n",
       " (4, 'ubiquitin-protein transferase activity'),\n",
       " (5, 'histone acetyltransferase activity'),\n",
       " (6, 'protease binding'),\n",
       " (7, 'single-stranded DNA binding'),\n",
       " (8, 'protein binding'),\n",
       " (9, 'negative regulation of transcription, DNA-templated'),\n",
       " (10, 'positive regulation of transcription by RNA polymerase II'),\n",
       " (11, 'promoter-specific chromatin binding'),\n",
       " (12, 'circadian behavior'),\n",
       " (13, 'cell adhesion')]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_GO_terms)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 1),\n",
       " (672, 2),\n",
       " (672, 3),\n",
       " (672, 4),\n",
       " (675, 5),\n",
       " (675, 6),\n",
       " (675, 7),\n",
       " (675, 8),\n",
       " (7157, 9),\n",
       " (7157, 10),\n",
       " (7157, 11),\n",
       " (7157, 12),\n",
       " (10296, 13),\n",
       " (10296, 8),\n",
       " (10296, 11),\n",
       " (7040, 13),\n",
       " (7040, 8),\n",
       " (64926, 13),\n",
       " (64926, 8),\n",
       " (64926, 6)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_gene_GO)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if all is well commit otherwise rollback, correct the issue in the code - rerun the code and if all well commit\n",
    "\n",
    "# connection.commit()\n",
    "connection.rollback()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x7feacad5dce0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## OPTION 2 - make lists and insert all data at the end\n",
    "\n",
    "# before we can do that we need to clear the table data if we did not rollback option 1\n",
    "\n",
    "delete_genes = \"DELETE FROM genes;\"\n",
    "delete_GO_terms = \"DELETE FROM GO_terms;\"\n",
    "delete_gene_GO = \"DELETE FROM gene_GO_term;\"\n",
    "\n",
    "cursor.execute(delete_genes)\n",
    "cursor.execute(delete_GO_terms)\n",
    "cursor.execute(delete_gene_GO)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "[]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "# check all 3 tables are empty\n",
    "\n",
    "cursor.execute(select_genes)\n",
    "print(cursor.fetchall())\n",
    "\n",
    "cursor.execute(select_GO_terms)\n",
    "print(cursor.fetchall())\n",
    "\n",
    "cursor.execute(select_gene_GO)\n",
    "print(cursor.fetchall())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commit the change\n",
    "\n",
    "connection.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "GO_terms_dict = {}\n",
    "GO_term_index = 0\n",
    "\n",
    "genes_rows = []\n",
    "GO_terms_rows = []\n",
    "gene_GO_rows = []\n",
    "\n",
    "with open(file_name) as genes_info:\n",
    "    # read the header line and ignore it\n",
    "    header_line = genes_info.readline() \n",
    "    \n",
    "    # read the rest of the lines and process them\n",
    "    for line in genes_info: \n",
    "        # break the line by tab and take the information needed in the table\n",
    "        line_list = line.strip().split(\"\\t\")\n",
    "        \n",
    "        # we put the gene id in a variable because we need it in gene_GO_term table\n",
    "        gene_id = line_list[0]\n",
    "        \n",
    "        # we take the gene id, symbol, description, \n",
    "        # which are the first 3 elements in the list - see header_line to confirm\n",
    "        # this will be a row in the genes table\n",
    "        gene_row = line_list[:3]\n",
    "        genes_rows.append(gene_row)\n",
    "        \n",
    "        # we take the go terms which are last in the list\n",
    "        # we breake them up into a list - they are separated by \";\"\n",
    "        GO_terms_list = line_list[-1].split(\";\")\n",
    "        \n",
    "        # we go through the list of go terms to create rows for the GO_terms table\n",
    "        # we need to keep track of the GO terms so we do not add them multiple times in the table\n",
    "        # one GO term can be associated with multiple genes and one genes can be associated with multiple GO terms\n",
    "        # a dictionary has unique keys and can store values foe each key\n",
    "        # we will use that to keep track of our GO terms\n",
    "        for GO_term in GO_terms_list:\n",
    "            # remove whitespace from the beginning and end of the GO_term\n",
    "            GO_term_name = GO_term.strip()\n",
    "            \n",
    "            # if we did not see this GO term before we add it to the dictionary and create an ID for it\n",
    "            # otherwise we get the id of the term from the dictionary\n",
    "            if GO_term_name not in GO_terms_dict:\n",
    "                GO_term_index += 1\n",
    "                GO_term_id = GO_term_index\n",
    "                GO_terms_dict[GO_term_name] = GO_term_index\n",
    "                # we will add this new GO term data to the GO_term table so prepare a row\n",
    "                GO_term_row = (GO_term_id, GO_term_name)\n",
    "                GO_terms_rows.append(GO_term_row)\n",
    "\n",
    "            else:\n",
    "                GO_term_id = GO_terms_dict[GO_term_name]\n",
    "            \n",
    "            # we add the connection between the gene and the GO term to the gene_GO_term table\n",
    "            # we prepare a row\n",
    "            gene_GO_row = (gene_id, GO_term_id)\n",
    "            gene_GO_rows.append(gene_GO_row)\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['672', 'BRCA1', 'BRCA1 DNA repair associated'],\n",
       " ['675', 'BRCA2', 'BRCA2 DNA repair associated'],\n",
       " ['7157', 'TP53', 'tumor protein p53'],\n",
       " ['10296', 'MAEA', 'macrophage erythroblast attacher, E3 ubiquitin ligase'],\n",
       " ['7040', 'TGFB1', 'transforming growth factor beta 1'],\n",
       " ['64926', 'RASAL3', 'RAS protein activator like 3']]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the lists\n",
    "\n",
    "genes_rows\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1, 'intrinsic apoptotic signaling pathway in response to DNA damage'),\n",
       " (2, 'transcription cis-regulatory region binding'),\n",
       " (3, 'transcription coactivator activity'),\n",
       " (4, 'ubiquitin-protein transferase activity'),\n",
       " (5, 'histone acetyltransferase activity'),\n",
       " (6, 'protease binding'),\n",
       " (7, 'single-stranded DNA binding'),\n",
       " (8, 'protein binding'),\n",
       " (9, 'negative regulation of transcription, DNA-templated'),\n",
       " (10, 'positive regulation of transcription by RNA polymerase II'),\n",
       " (11, 'promoter-specific chromatin binding'),\n",
       " (12, 'circadian behavior'),\n",
       " (13, 'cell adhesion')]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GO_terms_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('672', 1),\n",
       " ('672', 2),\n",
       " ('672', 3),\n",
       " ('672', 4),\n",
       " ('675', 5),\n",
       " ('675', 6),\n",
       " ('675', 7),\n",
       " ('675', 8),\n",
       " ('7157', 9),\n",
       " ('7157', 10),\n",
       " ('7157', 11),\n",
       " ('7157', 12),\n",
       " ('10296', 13),\n",
       " ('10296', 8),\n",
       " ('10296', 11),\n",
       " ('7040', 13),\n",
       " ('7040', 8),\n",
       " ('64926', 13),\n",
       " ('64926', 8),\n",
       " ('64926', 6)]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_GO_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x7feacad5dce0>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# insert the data from the lists \n",
    "# because we have many rows in the lists we use execute many\n",
    "\n",
    "cursor.executemany(insert_gene, genes_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x7feacad5dce0>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.executemany(insert_GO_term, GO_terms_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x7feacad5dce0>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.executemany(insert_gene_GO_term, gene_GO_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 'BRCA1', 'BRCA1 DNA repair associated'),\n",
       " (675, 'BRCA2', 'BRCA2 DNA repair associated'),\n",
       " (7040, 'TGFB1', 'transforming growth factor beta 1'),\n",
       " (7157, 'TP53', 'tumor protein p53'),\n",
       " (10296, 'MAEA', 'macrophage erythroblast attacher, E3 ubiquitin ligase'),\n",
       " (64926, 'RASAL3', 'RAS protein activator like 3')]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the data\n",
    "\n",
    "cursor.execute(select_genes)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(1, 'intrinsic apoptotic signaling pathway in response to DNA damage'),\n",
       " (2, 'transcription cis-regulatory region binding'),\n",
       " (3, 'transcription coactivator activity'),\n",
       " (4, 'ubiquitin-protein transferase activity'),\n",
       " (5, 'histone acetyltransferase activity'),\n",
       " (6, 'protease binding'),\n",
       " (7, 'single-stranded DNA binding'),\n",
       " (8, 'protein binding'),\n",
       " (9, 'negative regulation of transcription, DNA-templated'),\n",
       " (10, 'positive regulation of transcription by RNA polymerase II'),\n",
       " (11, 'promoter-specific chromatin binding'),\n",
       " (12, 'circadian behavior'),\n",
       " (13, 'cell adhesion')]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_GO_terms)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 1),\n",
       " (672, 2),\n",
       " (672, 3),\n",
       " (672, 4),\n",
       " (675, 5),\n",
       " (675, 6),\n",
       " (675, 7),\n",
       " (675, 8),\n",
       " (7157, 9),\n",
       " (7157, 10),\n",
       " (7157, 11),\n",
       " (7157, 12),\n",
       " (10296, 13),\n",
       " (10296, 8),\n",
       " (10296, 11),\n",
       " (7040, 13),\n",
       " (7040, 8),\n",
       " (64926, 13),\n",
       " (64926, 8),\n",
       " (64926, 6)]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_gene_GO)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commit the changes\n",
    "connection.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x7feacad5dce0>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## OPTION 3 - use the autoincrement for the GO_term id since it is not provided to us in the file\n",
    "\n",
    "# before we can do that we need to clear the table data\n",
    "delete_genes = \"DELETE FROM genes;\"\n",
    "delete_GO_terms = \"DELETE FROM GO_terms;\"\n",
    "delete_gene_GO = \"DELETE FROM gene_GO_term;\"\n",
    "\n",
    "cursor.execute(delete_genes)\n",
    "cursor.execute(delete_GO_terms)\n",
    "cursor.execute(delete_gene_GO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commit the change\n",
    "connection.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the insert statement for the GO_term table changed since we do not need to provide the id\n",
    "insert_GO_term1 = \"INSERT INTO GO_terms (name) VALUES (?);\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "GO_terms_dict = {}\n",
    "\n",
    "with open(file_name) as genes_info:\n",
    "    header_line = genes_info.readline() \n",
    "    for line in genes_info: \n",
    "        line_list = line.strip().split(\"\\t\")\n",
    "        gene_id = line_list[0]\n",
    "\n",
    "        gene_row = line_list[:3]\n",
    "        cursor.execute(insert_gene, gene_row)\n",
    "\n",
    "        GO_terms_list = line_list[-1].split(\";\")\n",
    "        \n",
    "        for GO_term in GO_terms_list:\n",
    "            GO_term_name = GO_term.strip()\n",
    "            if GO_term_name not in GO_terms_dict:\n",
    "                cursor.execute(insert_GO_term1, [GO_term_name])\n",
    "                \n",
    "                # get the autogenerated id from the database after the insert\n",
    "                GO_term_id = cursor.lastrowid\n",
    "                \n",
    "                GO_terms_dict[GO_term_name] = GO_term_id\n",
    "\n",
    "            else:\n",
    "                GO_term_id = GO_terms_dict[GO_term_name]\n",
    "\n",
    "            gene_GO_row = (gene_id, GO_term_id)\n",
    "            cursor.execute(insert_gene_GO_term, gene_GO_row)\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 'BRCA1', 'BRCA1 DNA repair associated'),\n",
       " (675, 'BRCA2', 'BRCA2 DNA repair associated'),\n",
       " (7040, 'TGFB1', 'transforming growth factor beta 1'),\n",
       " (7157, 'TP53', 'tumor protein p53'),\n",
       " (10296, 'MAEA', 'macrophage erythroblast attacher, E3 ubiquitin ligase'),\n",
       " (64926, 'RASAL3', 'RAS protein activator like 3')]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the data\n",
    "\n",
    "cursor.execute(select_genes)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(14, 'intrinsic apoptotic signaling pathway in response to DNA damage'),\n",
       " (15, 'transcription cis-regulatory region binding'),\n",
       " (16, 'transcription coactivator activity'),\n",
       " (17, 'ubiquitin-protein transferase activity'),\n",
       " (18, 'histone acetyltransferase activity'),\n",
       " (19, 'protease binding'),\n",
       " (20, 'single-stranded DNA binding'),\n",
       " (21, 'protein binding'),\n",
       " (22, 'negative regulation of transcription, DNA-templated'),\n",
       " (23, 'positive regulation of transcription by RNA polymerase II'),\n",
       " (24, 'promoter-specific chromatin binding'),\n",
       " (25, 'circadian behavior'),\n",
       " (26, 'cell adhesion')]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_GO_terms)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(672, 14),\n",
       " (672, 15),\n",
       " (672, 16),\n",
       " (672, 17),\n",
       " (675, 18),\n",
       " (675, 19),\n",
       " (675, 20),\n",
       " (675, 21),\n",
       " (7157, 22),\n",
       " (7157, 23),\n",
       " (7157, 24),\n",
       " (7157, 25),\n",
       " (10296, 26),\n",
       " (10296, 21),\n",
       " (10296, 24),\n",
       " (7040, 26),\n",
       " (7040, 21),\n",
       " (64926, 26),\n",
       " (64926, 21),\n",
       " (64926, 19)]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cursor.execute(select_gene_GO)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# commit the changes\n",
    "connection.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "#### Query a database \n",
    "\n",
    "How many genes we have?  \n",
    "How many go terms we have?   \n",
    "How many go terms are associated which each gene?   \n",
    "How many genes are associated with each GO term?   \n",
    "Select all pairs of gene symbol and go term description that have a record in the gene_GO_term table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "# Bring in the functions from class for a nice display of the results\n",
    "\n",
    "def get_header(cursor):\n",
    "    '''\n",
    "    Makes a tab delimited header row from the cursor description.\n",
    "    Arguments:\n",
    "        cursor: a cursor after a select query\n",
    "    Returns:\n",
    "        string: A string consisting of the column names separated by tabs, no new line\n",
    "    '''\n",
    "    return '\\t'.join([row[0] for row in cursor.description])\n",
    "\n",
    "def get_results(cursor):\n",
    "    '''\n",
    "    Makes a tab delimited table from the cursor results.\n",
    "    Arguments:\n",
    "        cursor: a cursor after a select query\n",
    "    Returns:\n",
    "        string: A string consisting of the column names separated by tabs, no new line\n",
    "    ''' \n",
    "    res = list()\n",
    "    for row in cursor.fetchall():        \n",
    "        res.append('\\t'.join(list(map(str,row))))\n",
    "    return \"\\n\".join(res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(6,)]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# query1: How many genes we have?\n",
    "\n",
    "query1 = \"SELECT count(gene_id) 'Number of genes' FROM genes;\"\n",
    "\n",
    "cursor.execute(query1)\n",
    "result = cursor.fetchall()\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# to get the number we index the list and the tuple\n",
    "\n",
    "result[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "cursor.execute(query1)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of GO terms\n",
      "13\n"
     ]
    }
   ],
   "source": [
    "# query2: How many GO terms we have?\n",
    "\n",
    "query2 = \"SELECT count(*) 'Number of GO terms' FROM GO_terms;\"\n",
    "\n",
    "cursor.execute(query2)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_id\tNumber of GO terms\n",
      "672\t4\n",
      "675\t4\n",
      "7040\t2\n",
      "7157\t4\n",
      "10296\t3\n",
      "64926\t3\n"
     ]
    }
   ],
   "source": [
    "# query3: How many go terms are associated which each gene?   \n",
    "\n",
    "query3 = '''\n",
    "SELECT gene_id, count(GO_term_id) 'Number of GO terms' \n",
    "FROM gene_GO_term\n",
    "GROUP BY gene_id;\n",
    "'''\n",
    "\n",
    "cursor.execute(query3)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tNumber of GO terms\n",
      "BRCA1\t4\n",
      "BRCA2\t4\n",
      "MAEA\t3\n",
      "RASAL3\t3\n",
      "TGFB1\t2\n",
      "TP53\t4\n"
     ]
    }
   ],
   "source": [
    "# if we want the gene_symbol\n",
    "\n",
    "query3 = '''\n",
    "SELECT gene_symbol Gene, count(gt.GO_term_id) 'Number of GO terms' \n",
    "FROM gene_GO_term gt\n",
    "    JOIN genes g ON g.gene_id = gt.gene_id\n",
    "GROUP BY gene_symbol;\n",
    "'''\n",
    "\n",
    "cursor.execute(query3)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO_term_id\tNumber of genes\n",
      "14\t1\n",
      "15\t1\n",
      "16\t1\n",
      "17\t1\n",
      "18\t1\n",
      "19\t2\n",
      "20\t1\n",
      "21\t4\n",
      "22\t1\n",
      "23\t1\n",
      "24\t2\n",
      "25\t1\n",
      "26\t3\n"
     ]
    }
   ],
   "source": [
    "# query4: How many genes are associated with each GO term?  \n",
    "\n",
    "query4 = '''\n",
    "SELECT GO_term_id, count(gene_id) 'Number of genes' \n",
    "FROM gene_GO_term\n",
    "GROUP BY GO_term_id;\n",
    "'''\n",
    "\n",
    "cursor.execute(query4)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO term</th>\n",
       "      <th>Number of genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cell adhesion</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>circadian behavior</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>histone acetyltransferase activity</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intrinsic apoptotic signaling pathway in respo...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>negative regulation of transcription, DNA-temp...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>positive regulation of transcription by RNA po...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>promoter-specific chromatin binding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>protease binding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>protein binding</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>single-stranded DNA binding</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>transcription cis-regulatory region binding</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>transcription coactivator activity</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ubiquitin-protein transferase activity</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GO term  Number of genes\n",
       "0                                       cell adhesion                3\n",
       "1                                  circadian behavior                1\n",
       "2                  histone acetyltransferase activity                1\n",
       "3   intrinsic apoptotic signaling pathway in respo...                1\n",
       "4   negative regulation of transcription, DNA-temp...                1\n",
       "5   positive regulation of transcription by RNA po...                1\n",
       "6                 promoter-specific chromatin binding                2\n",
       "7                                    protease binding                2\n",
       "8                                     protein binding                4\n",
       "9                         single-stranded DNA binding                1\n",
       "10        transcription cis-regulatory region binding                1\n",
       "11                 transcription coactivator activity                1\n",
       "12             ubiquitin-protein transferase activity                1"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if we want the GO term name\n",
    "\n",
    "query4 = '''\n",
    "SELECT name 'GO term', count(gene_id) 'Number of genes' \n",
    "FROM gene_GO_term gt\n",
    "    JOIN GO_terms t ON t.GO_term_id = gt.GO_term_id\n",
    "GROUP BY name;\n",
    "'''\n",
    "\n",
    "cursor.execute(query4)\n",
    "\n",
    "# create a dataframe of the results\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tGO term\n",
      "BRCA1\tintrinsic apoptotic signaling pathway in response to DNA damage\n",
      "BRCA1\ttranscription cis-regulatory region binding\n",
      "BRCA1\ttranscription coactivator activity\n",
      "BRCA1\tubiquitin-protein transferase activity\n",
      "BRCA2\thistone acetyltransferase activity\n",
      "BRCA2\tprotease binding\n",
      "BRCA2\tsingle-stranded DNA binding\n",
      "BRCA2\tprotein binding\n",
      "TP53\tnegative regulation of transcription, DNA-templated\n",
      "TP53\tpositive regulation of transcription by RNA polymerase II\n",
      "TP53\tpromoter-specific chromatin binding\n",
      "TP53\tcircadian behavior\n",
      "MAEA\tcell adhesion\n",
      "MAEA\tprotein binding\n",
      "MAEA\tpromoter-specific chromatin binding\n",
      "TGFB1\tcell adhesion\n",
      "TGFB1\tprotein binding\n",
      "RASAL3\tcell adhesion\n",
      "RASAL3\tprotein binding\n",
      "RASAL3\tprotease binding\n"
     ]
    }
   ],
   "source": [
    "# query5: Select all pairs of gene symbol and go term description that have a record in the gene_GO_term table.\n",
    "\n",
    "query5 = '''\n",
    "SELECT gene_symbol Gene, name 'GO term' \n",
    "FROM gene_GO_term gt\n",
    "    JOIN GO_terms t ON t.GO_term_id = gt.GO_term_id\n",
    "    JOIN genes g on g.gene_id = gt.gene_id;\n",
    "'''\n",
    "\n",
    "cursor.execute(query5)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we are done working with this database\n",
    "# close cursor and connection\n",
    "\n",
    "cursor.close()\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "#### Query a database \n",
    "Download the database from the following link:\n",
    "https://www.sqlitetutorial.net/sqlite-sample-database/\n",
    "\n",
    "<img src = https://www.sqlitetutorial.net/wp-content/uploads/2015/11/sqlite-sample-database-color.jpg width = 675>\n",
    "\n",
    "https://www.sqlitetutorial.net/wp-content/uploads/2015/11/sqlite-sample-database-color.jpg\n",
    "\n",
    "Retrieve how many tracks are associated with each artist.    \n",
    "What is the overall mean number of tracks per album?    \n",
    "For a given artist, retrieve all the albums (name), and tracks (name), they are associated with.   \n",
    "\n",
    "Feel free to build other quesries for questions you may have.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "# connect to the database\n",
    "\n",
    "connection = connect(\"chinook.db\")\n",
    "cursor = connection.cursor()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>albums</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sqlite_sequence</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>artists</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>customers</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>employees</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>genres</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>invoices</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>invoice_items</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>media_types</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>playlists</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>playlist_track</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>sqlite_autoindex_playlist_track_1</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tracks</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IFK_AlbumArtistId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IFK_CustomerSupportRepId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>IFK_EmployeeReportsTo</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IFK_InvoiceCustomerId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>IFK_InvoiceLineInvoiceId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IFK_InvoiceLineTrackId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IFK_PlaylistTrackTrackId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IFK_TrackAlbumId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>IFK_TrackGenreId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IFK_TrackMediaTypeId</td>\n",
       "      <td>index</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>sqlite_stat1</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 name   type\n",
       "0                              albums  table\n",
       "1                     sqlite_sequence  table\n",
       "2                             artists  table\n",
       "3                           customers  table\n",
       "4                           employees  table\n",
       "5                              genres  table\n",
       "6                            invoices  table\n",
       "7                       invoice_items  table\n",
       "8                         media_types  table\n",
       "9                           playlists  table\n",
       "10                     playlist_track  table\n",
       "11  sqlite_autoindex_playlist_track_1  index\n",
       "12                             tracks  table\n",
       "13                  IFK_AlbumArtistId  index\n",
       "14           IFK_CustomerSupportRepId  index\n",
       "15              IFK_EmployeeReportsTo  index\n",
       "16              IFK_InvoiceCustomerId  index\n",
       "17           IFK_InvoiceLineInvoiceId  index\n",
       "18             IFK_InvoiceLineTrackId  index\n",
       "19           IFK_PlaylistTrackTrackId  index\n",
       "20                   IFK_TrackAlbumId  index\n",
       "21                   IFK_TrackGenreId  index\n",
       "22               IFK_TrackMediaTypeId  index\n",
       "23                       sqlite_stat1  table"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the tables in the database\n",
    "\n",
    "sel_tables = \"select name, type from sqlite_master\" \n",
    "cursor.execute(sel_tables)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ArtistId</th>\n",
       "      <th>count(trackid)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>271</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>272</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>273</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>274</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>275</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ArtistId  count(trackid)\n",
       "0           1              18\n",
       "1           2               4\n",
       "2           3              15\n",
       "3           4              13\n",
       "4           5              12\n",
       "..        ...             ...\n",
       "199       271               1\n",
       "200       272               1\n",
       "201       273               1\n",
       "202       274               1\n",
       "203       275               1\n",
       "\n",
       "[204 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# task1: Retrieve how many tracks are associated with each artist.\n",
    "\n",
    "task1 = '''\n",
    "SELECT artistid, count(trackid) \n",
    "FROM tracks t \n",
    "   JOIN albums ab ON t.albumid = ab.albumid \n",
    "GROUP BY artistid\n",
    "'''\n",
    "\n",
    "cursor.execute(task1)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>No of tracks</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>213</td>\n",
       "      <td>Iron Maiden</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>135</td>\n",
       "      <td>U2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>114</td>\n",
       "      <td>Led Zeppelin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>112</td>\n",
       "      <td>Metallica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>92</td>\n",
       "      <td>Lost</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>1</td>\n",
       "      <td>Academy of St. Martin in the Fields, Sir Nevil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>1</td>\n",
       "      <td>Academy of St. Martin in the Fields, John Birc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>1</td>\n",
       "      <td>Academy of St. Martin in the Fields Chamber En...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>1</td>\n",
       "      <td>Aaron Goldberg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>1</td>\n",
       "      <td>Aaron Copland &amp; London Symphony Orchestra</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     No of tracks                                               Name\n",
       "0             213                                        Iron Maiden\n",
       "1             135                                                 U2\n",
       "2             114                                       Led Zeppelin\n",
       "3             112                                          Metallica\n",
       "4              92                                               Lost\n",
       "..            ...                                                ...\n",
       "199             1  Academy of St. Martin in the Fields, Sir Nevil...\n",
       "200             1  Academy of St. Martin in the Fields, John Birc...\n",
       "201             1  Academy of St. Martin in the Fields Chamber En...\n",
       "202             1                                     Aaron Goldberg\n",
       "203             1          Aaron Copland & London Symphony Orchestra\n",
       "\n",
       "[204 rows x 2 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if we want the srtist name and ordered data\n",
    "\n",
    "task1 = '''\n",
    "SELECT  count(trackid) 'No of tracks', a.name\n",
    "FROM tracks t \n",
    "   JOIN albums ab ON t.albumid = ab.albumid \n",
    "   JOIN artists a ON ab.artistid = a.artistid\n",
    "GROUP BY a.name\n",
    "ORDER BY count(trackid) DESC\n",
    "'''\n",
    "\n",
    "cursor.execute(task1)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tracks per album</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Tracks per album\n",
       "0                  57\n",
       "1                  34\n",
       "2                  30\n",
       "3                  26\n",
       "4                  25\n",
       "..                ...\n",
       "342                 1\n",
       "343                 1\n",
       "344                 1\n",
       "345                 1\n",
       "346                 1\n",
       "\n",
       "[347 rows x 1 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# task2: What is the overall mean number of tracks per album?\n",
    "\n",
    "# first let's get the number of Tracks per album before we do the mean\n",
    "task2 = '''\n",
    "SELECT count(trackid) 'Tracks per album'\n",
    "FROM tracks t \n",
    "   JOIN albums ab ON t.albumid = ab.albumid \n",
    "GROUP BY ab.albumid\n",
    "ORDER BY count(trackid) DESC\n",
    "'''\n",
    "\n",
    "cursor.execute(task2)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the list of aggregate functions - we see that for mean we use avg()  \n",
    "https://www.sqlite.org/lang_aggfunc.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean number of tracks per album\n",
      "10.095100864553315\n"
     ]
    }
   ],
   "source": [
    "# let's do the mean of the results from the previous select \n",
    "# a select returns data in a table format so we can select from the result of a select\n",
    "\n",
    "task2 = '''\n",
    "SELECT avg(c) 'Mean number of tracks per album'\n",
    "FROM\n",
    "    (SELECT count(trackid) c\n",
    "    FROM tracks t \n",
    "       JOIN albums ab ON t.albumid = ab.albumid \n",
    "    GROUP BY ab.albumid)\n",
    "'''\n",
    "\n",
    "cursor.execute(task2)\n",
    "\n",
    "print(get_header(cursor))\n",
    "print(get_results(cursor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Title</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>...And Justice For All</td>\n",
       "      <td>...And Justice For All</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>...And Justice For All</td>\n",
       "      <td>Blackened</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>...And Justice For All</td>\n",
       "      <td>Dyers Eve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>...And Justice For All</td>\n",
       "      <td>Eye Of The Beholder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>...And Justice For All</td>\n",
       "      <td>Harvester Of Sorrow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>St. Anger</td>\n",
       "      <td>Shoot Me Again</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>St. Anger</td>\n",
       "      <td>Some Kind Of Monster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>St. Anger</td>\n",
       "      <td>St. Anger</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>St. Anger</td>\n",
       "      <td>Sweet Amber</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Metallica</td>\n",
       "      <td>St. Anger</td>\n",
       "      <td>The Unnamed Feeling</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Name                   Title                    Name\n",
       "0    Metallica  ...And Justice For All  ...And Justice For All\n",
       "1    Metallica  ...And Justice For All               Blackened\n",
       "2    Metallica  ...And Justice For All               Dyers Eve\n",
       "3    Metallica  ...And Justice For All     Eye Of The Beholder\n",
       "4    Metallica  ...And Justice For All     Harvester Of Sorrow\n",
       "..         ...                     ...                     ...\n",
       "107  Metallica               St. Anger          Shoot Me Again\n",
       "108  Metallica               St. Anger    Some Kind Of Monster\n",
       "109  Metallica               St. Anger               St. Anger\n",
       "110  Metallica               St. Anger             Sweet Amber\n",
       "111  Metallica               St. Anger     The Unnamed Feeling\n",
       "\n",
       "[112 rows x 3 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# task3: For a given artist, retrieve all the albums (name), \n",
    "# and tracks (name), they are associated with.\n",
    "\n",
    "task3 = '''\n",
    "SELECT  a.name, title, t.name\n",
    "FROM tracks t \n",
    "   JOIN albums ab ON t.albumid = ab.albumid \n",
    "   JOIN artists a ON ab.artistid = a.artistid\n",
    "WHERE a.name = \"Metallica\"\n",
    "ORDER BY title, t.name\n",
    "'''\n",
    "\n",
    "cursor.execute(task3)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Title</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Breaking The Rules</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>C.O.D.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Evil Walks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>For Those About To Rock (We Salute You)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Inject The Venom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Let's Get It Up</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Night Of The Long Knives</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Put The Finger On You</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Snowballed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>For Those About To Rock We Salute You</td>\n",
       "      <td>Spellbound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Bad Boy Boogie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Dog Eat Dog</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Go Down</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Hell Ain't A Bad Place To Be</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Overdose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Problem Child</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AC/DC</td>\n",
       "      <td>Let There Be Rock</td>\n",
       "      <td>Whole Lotta Rosie</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name                                  Title  \\\n",
       "0   AC/DC  For Those About To Rock We Salute You   \n",
       "1   AC/DC  For Those About To Rock We Salute You   \n",
       "2   AC/DC  For Those About To Rock We Salute You   \n",
       "3   AC/DC  For Those About To Rock We Salute You   \n",
       "4   AC/DC  For Those About To Rock We Salute You   \n",
       "5   AC/DC  For Those About To Rock We Salute You   \n",
       "6   AC/DC  For Those About To Rock We Salute You   \n",
       "7   AC/DC  For Those About To Rock We Salute You   \n",
       "8   AC/DC  For Those About To Rock We Salute You   \n",
       "9   AC/DC  For Those About To Rock We Salute You   \n",
       "10  AC/DC                      Let There Be Rock   \n",
       "11  AC/DC                      Let There Be Rock   \n",
       "12  AC/DC                      Let There Be Rock   \n",
       "13  AC/DC                      Let There Be Rock   \n",
       "14  AC/DC                      Let There Be Rock   \n",
       "15  AC/DC                      Let There Be Rock   \n",
       "16  AC/DC                      Let There Be Rock   \n",
       "17  AC/DC                      Let There Be Rock   \n",
       "\n",
       "                                       Name  \n",
       "0                        Breaking The Rules  \n",
       "1                                    C.O.D.  \n",
       "2                                Evil Walks  \n",
       "3   For Those About To Rock (We Salute You)  \n",
       "4                          Inject The Venom  \n",
       "5                           Let's Get It Up  \n",
       "6                  Night Of The Long Knives  \n",
       "7                     Put The Finger On You  \n",
       "8                                Snowballed  \n",
       "9                                Spellbound  \n",
       "10                           Bad Boy Boogie  \n",
       "11                              Dog Eat Dog  \n",
       "12                                  Go Down  \n",
       "13             Hell Ain't A Bad Place To Be  \n",
       "14                        Let There Be Rock  \n",
       "15                                 Overdose  \n",
       "16                            Problem Child  \n",
       "17                        Whole Lotta Rosie  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if we want to make a generic select where we supply the artist name\n",
    "task3 = '''\n",
    "SELECT  a.name, title, t.name\n",
    "FROM tracks t \n",
    "   JOIN albums ab ON t.albumid = ab.albumid \n",
    "   JOIN artists a ON ab.artistid = a.artistid\n",
    "WHERE a.name = ?\n",
    "ORDER BY title, t.name\n",
    "'''\n",
    "\n",
    "cursor.execute(task3, [\"AC/DC\"])\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [tp[0] for tp in cursor.description]\n",
    "res_df = pd.DataFrame(data, columns = header)\n",
    "res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# answer more interesting questions here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# when we are done working with the database\n",
    "# close the cursor and the connection\n",
    "\n",
    "cursor.close()\n",
    "connection.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "#### Variant database\n",
    "\n",
    "Create a variant sqlite database with 2 tables variant and var_info from the file `variant_file.vcf`.    \n",
    "The header of the file is:\n",
    "`#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tGroup_A/IP2-50/tophat_out/IP2-50_accepted_hits.bam`\n",
    "\n",
    "The first 6 columns from the file will be columns in the `variant` table. The `variant` table will also have a `variant_id` - autoincrement column to uniquely identify the variant in this database.    \n",
    "\n",
    "\n",
    "The `var_info` table will contain the data from the INFO column in the file.  The `var_info` table will have the following columns: `variant_id`, `tag` and `value`. \n",
    "For each variant, the pairs tag=values are separated by \";\" and the values are separated by \",\".  Each entry in the var_info table will have a unique variant_id, tag and value column combination.\n",
    "\n",
    "\n",
    "Run the following querries:\n",
    "- A select that retrieves how many variants we have.\n",
    "- A select of the variant(s) with the highest number associated tag-value pairs.\n",
    "- A select of INDEL variants with DP values > 150.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "# create and connect to the variant database\n",
    "# this should be a new file\n",
    "# make sure it does not exist in your current folder\n",
    "\n",
    "connection = connect(\"variant.sqlite\")\n",
    "\n",
    "# create a cursor\n",
    "cursor = connection.cursor()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create tables in the database\n",
    "\n",
    "# write the create statements\n",
    "\n",
    "#-------\n",
    "create_variant = '''\n",
    "CREATE TABLE IF NOT EXISTS variant (\n",
    "      variant_id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "      CHROM TEXT NOT NULL,\n",
    "      POS BIGINT  NOT NULL,\n",
    "      ID TEXT,\n",
    "      REF TEXT, \n",
    "      ALT TEXT, \n",
    "      QUAL TEXT\n",
    "    );\n",
    "'''\n",
    "#-------\n",
    "create_var_info = '''\n",
    "CREATE TABLE IF NOT EXISTS var_info (\n",
    "      variant_id INTEGER NOT NULL,\n",
    "      tag INTEGER NOT NULL,\n",
    "      value INTEGER NOT NULL\n",
    "    );\n",
    "'''\n",
    "#-------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we do the same thing multiple times - we can use a function\n",
    "\n",
    "def run_create_table(sql, cursor, connection):\n",
    "    try:\n",
    "        cursor.execute(sql)\n",
    "    except connection.DatabaseError:\n",
    "        print(\"Creating the table resulted in a database error!\")\n",
    "        connection.rollback()\n",
    "        raise\n",
    "    else:\n",
    "        connection.commit()\n",
    "    finally:\n",
    "        print(\"done!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sqlite_master table - should be empty at this point\n",
    "\n",
    "select_master = \"SELECT name, type FROM sqlite_master;\"\n",
    "cursor.execute(select_master)\n",
    "cursor.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "# run the commands\n",
    "run_create_table(create_variant, cursor, connection)\n",
    "run_create_table(create_var_info, cursor, connection)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>variant</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sqlite_sequence</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>var_info</td>\n",
       "      <td>table</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name   type\n",
       "0          variant  table\n",
       "1  sqlite_sequence  table\n",
       "2         var_info  table"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the sqlite_master table \n",
    "# should have the two tables\n",
    "\n",
    "select_master = \"SELECT name, type FROM sqlite_master;\"\n",
    "cursor.execute(select_master)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the insert statements\n",
    "\n",
    "#-------\n",
    "insert_variant = \"\"\"\n",
    "INSERT INTO variant (CHROM, POS, ID, REF, ALT, QUAL) VALUES (?,?,?,?,?,?)\n",
    "\"\"\"\n",
    "#-------\n",
    "\n",
    "insert_info = \"\"\"\n",
    "INSERT INTO var_info VALUES (?,?,?)\n",
    "\"\"\"\n",
    "#-------\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse the file and add data to the two tables\n",
    "\n",
    "# open file for reading\n",
    "with open(\"variant_file.vcf\") as vf:\n",
    "    # go through file lines\n",
    "    for line in vf:\n",
    "        # skip comment lines\n",
    "        if not line.startswith(\"#\"):\n",
    "            # remove new line at the end and break the line by tab\n",
    "            line_lst = line.strip().split(\"\\t\")\n",
    "            # take the first 6 elements - to go into a variant table row\n",
    "            var_vals = line_lst[:6]\n",
    "            # run the insert command\n",
    "            cursor.execute(insert_variant, var_vals)\n",
    "            # take the variant id from the cursor attribute lastrowid\n",
    "            variant_id = cursor.lastrowid\n",
    "            # take the info column data 3rd from the end 8th from the beginning\n",
    "            # and break it by \";\"\n",
    "            info_lst = line_lst[-3].split(\";\")\n",
    "            # go though the list - it conains tag=values pairs\n",
    "            for pair in info_lst:\n",
    "                # break the pair by =\n",
    "                pair_lst = pair.split(\"=\")\n",
    "                # if the list has only one element \n",
    "                # add that as a tag with value 1 (True)\n",
    "                if len(pair_lst) == 1:\n",
    "                    var_info_vals = [variant_id, pair, 1]\n",
    "                    # execute insert for variant info\n",
    "                    cursor.execute(insert_info, var_info_vals)\n",
    "                else:\n",
    "                    # separate tag and values \n",
    "                    # split the vals by \",\"\n",
    "                    tag, vals = pair_lst\n",
    "                    for val in vals.split(\",\"):\n",
    "                        # execute insert for variant info\n",
    "                        # create the list of three values\n",
    "                        # use eval to evaluate the value to the proper datatype\n",
    "                        # i.e. make the val string a number \n",
    "                        var_info_vals = [variant_id, tag, eval(val)]\n",
    "                        # execute insert for variant info\n",
    "                        cursor.execute(insert_info, var_info_vals)\n",
    "                        \n",
    "    # commit/save the inserts/changes\n",
    "    connection.commit()\n",
    "                        \n",
    "\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>237763</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>6.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>564452</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>19.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>564464</td>\n",
       "      <td>.</td>\n",
       "      <td>GGG</td>\n",
       "      <td>GGGG</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>565286</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>36.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>567242</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193604</th>\n",
       "      <td>193605</td>\n",
       "      <td>Y</td>\n",
       "      <td>59025774</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>17.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193605</th>\n",
       "      <td>193606</td>\n",
       "      <td>Y</td>\n",
       "      <td>59025960</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193606</th>\n",
       "      <td>193607</td>\n",
       "      <td>Y</td>\n",
       "      <td>59027700</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>30.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193607</th>\n",
       "      <td>193608</td>\n",
       "      <td>Y</td>\n",
       "      <td>59027910</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>5.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193608</th>\n",
       "      <td>193609</td>\n",
       "      <td>Y</td>\n",
       "      <td>59028548</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>6.02</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193609 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant_id CHROM       POS ID  REF   ALT  QUAL\n",
       "0                1     1    237763  .    G     A  6.02\n",
       "1                2     1    564452  .    A     G  19.6\n",
       "2                3     1    564464  .  GGG  GGGG    20\n",
       "3                4     1    565286  .    C     T  36.1\n",
       "4                5     1    567242  .    G     A  18.8\n",
       "...            ...   ...       ... ..  ...   ...   ...\n",
       "193604      193605     Y  59025774  .    C     T  17.1\n",
       "193605      193606     Y  59025960  .    C     G  6.02\n",
       "193606      193607     Y  59027700  .    A     C  30.1\n",
       "193607      193608     Y  59027910  .    G     A  5.29\n",
       "193608      193609     Y  59028548  .    C     G  6.02\n",
       "\n",
       "[193609 rows x 7 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the variant table \n",
    "# should have data now\n",
    "\n",
    "select_variant = \"SELECT * FROM variant;\"\n",
    "cursor.execute(select_variant)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>DP</td>\n",
       "      <td>2.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>VDB</td>\n",
       "      <td>0.0261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>AF1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>AC1</td>\n",
       "      <td>2.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>DP4</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034307</th>\n",
       "      <td>193609</td>\n",
       "      <td>DP4</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034308</th>\n",
       "      <td>193609</td>\n",
       "      <td>DP4</td>\n",
       "      <td>2.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034309</th>\n",
       "      <td>193609</td>\n",
       "      <td>DP4</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034310</th>\n",
       "      <td>193609</td>\n",
       "      <td>MQ</td>\n",
       "      <td>20.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2034311</th>\n",
       "      <td>193609</td>\n",
       "      <td>FQ</td>\n",
       "      <td>-33.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2034312 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant_id  tag    value\n",
       "0                 1   DP   2.0000\n",
       "1                 1  VDB   0.0261\n",
       "2                 1  AF1   1.0000\n",
       "3                 1  AC1   2.0000\n",
       "4                 1  DP4   0.0000\n",
       "...             ...  ...      ...\n",
       "2034307      193609  DP4   0.0000\n",
       "2034308      193609  DP4   2.0000\n",
       "2034309      193609  DP4   0.0000\n",
       "2034310      193609   MQ  20.0000\n",
       "2034311      193609   FQ -33.0000\n",
       "\n",
       "[2034312 rows x 3 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the var_info table \n",
    "# should have data now\n",
    "\n",
    "select_var_info = \"SELECT * FROM var_info;\"\n",
    "cursor.execute(select_var_info)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the querries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Var. No.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>193609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Var. No.\n",
       "0    193609"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select that retrieves how many variants we have.\n",
    "\n",
    "# we use count\n",
    "\n",
    "select_var_count = \"SELECT count(variant_id) 'Var. No.' FROM variant;\"\n",
    "cursor.execute(select_var_count)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>InfoNo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>193584</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>193574</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>193513</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>193313</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>193303</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193604</th>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193605</th>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193606</th>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193607</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193608</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193609 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        variant_id  InfoNo\n",
       "0           193584      15\n",
       "1           193574      15\n",
       "2           193513      15\n",
       "3           193313      15\n",
       "4           193303      15\n",
       "...            ...     ...\n",
       "193604           9      10\n",
       "193605           8      10\n",
       "193606           4      10\n",
       "193607           2      10\n",
       "193608           1      10\n",
       "\n",
       "[193609 rows x 2 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select of the variant(s) with the highest number associated tag-value pairs.\n",
    "\n",
    "\n",
    "# count the number of tag-val pairs per variant \n",
    "# we use group by to group by variant_id in the var_info table\n",
    "# use order by to sort the data\n",
    "\n",
    "select_var_infocount = \"\"\"\n",
    "SELECT variant_id, count(*) InfoNo \n",
    "FROM var_info\n",
    "GROUP BY variant_id\n",
    "ORDER BY count(*)  DESC;\n",
    "\"\"\"\n",
    "cursor.execute(select_var_infocount)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>InfoNo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   InfoNo\n",
       "0      15"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of tag-val pairs per variant \n",
    "# we use group by to group by variant_id in the var_info table\n",
    "# use order by to sort the data\n",
    "\n",
    "# select the top count - order by desc and limit 1\n",
    "\n",
    "select_var_infocount = \"\"\"\n",
    "SELECT InfoNo\n",
    "FROM \n",
    "    (SELECT variant_id, count(*) InfoNo \n",
    "    FROM var_info\n",
    "    GROUP BY variant_id\n",
    "    ORDER BY count(*) DESC  \n",
    "    LIMIT 1);\n",
    "\"\"\"\n",
    "cursor.execute(select_var_infocount)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>max(InfoNo)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   max(InfoNo)\n",
       "0           15"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of tag-val pairs per variant \n",
    "# we use group by to group by variant_id in the var_info table\n",
    "# use order by to sort the data\n",
    "\n",
    "# select the max count \n",
    "\n",
    "select_var_infocount = \"\"\"\n",
    "SELECT max(InfoNo)\n",
    "FROM \n",
    "    (SELECT variant_id, count(*) InfoNo \n",
    "    FROM var_info\n",
    "    GROUP BY variant_id\n",
    "    ORDER BY count(*) DESC);\n",
    "\"\"\"\n",
    "cursor.execute(select_var_infocount)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>InfoNo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>193584</td>\n",
       "      <td>Y</td>\n",
       "      <td>59009514</td>\n",
       "      <td>.</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>21.5</td>\n",
       "      <td>193584</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>193574</td>\n",
       "      <td>Y</td>\n",
       "      <td>59001398</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCGG</td>\n",
       "      <td>3.43</td>\n",
       "      <td>193574</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>193513</td>\n",
       "      <td>Y</td>\n",
       "      <td>21152664</td>\n",
       "      <td>.</td>\n",
       "      <td>AAA</td>\n",
       "      <td>A</td>\n",
       "      <td>123</td>\n",
       "      <td>193513</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>193313</td>\n",
       "      <td>X</td>\n",
       "      <td>153590944</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCC</td>\n",
       "      <td>CCCCC</td>\n",
       "      <td>4.07</td>\n",
       "      <td>193313</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>193303</td>\n",
       "      <td>X</td>\n",
       "      <td>153577133</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TAG</td>\n",
       "      <td>117</td>\n",
       "      <td>193303</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1740</th>\n",
       "      <td>1214</td>\n",
       "      <td>1</td>\n",
       "      <td>17393706</td>\n",
       "      <td>.</td>\n",
       "      <td>AAAGAA</td>\n",
       "      <td>AAA</td>\n",
       "      <td>41.8</td>\n",
       "      <td>1214</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1741</th>\n",
       "      <td>962</td>\n",
       "      <td>1</td>\n",
       "      <td>14276752</td>\n",
       "      <td>.</td>\n",
       "      <td>GCC</td>\n",
       "      <td>GC</td>\n",
       "      <td>8.73</td>\n",
       "      <td>962</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1742</th>\n",
       "      <td>946</td>\n",
       "      <td>1</td>\n",
       "      <td>14106397</td>\n",
       "      <td>.</td>\n",
       "      <td>CCT</td>\n",
       "      <td>CCTCCT</td>\n",
       "      <td>53.9</td>\n",
       "      <td>946</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1743</th>\n",
       "      <td>225</td>\n",
       "      <td>1</td>\n",
       "      <td>3697549</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AA,AAA</td>\n",
       "      <td>24.9</td>\n",
       "      <td>225</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>760814</td>\n",
       "      <td>.</td>\n",
       "      <td>TTTCTTT</td>\n",
       "      <td>TTT</td>\n",
       "      <td>5.8</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1745 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id CHROM        POS ID      REF     ALT  QUAL  variant_id  \\\n",
       "0         193584     Y   59009514  .       GC       G  21.5      193584   \n",
       "1         193574     Y   59001398  .        G   GGCGG  3.43      193574   \n",
       "2         193513     Y   21152664  .      AAA       A   123      193513   \n",
       "3         193313     X  153590944  .     CCCC   CCCCC  4.07      193313   \n",
       "4         193303     X  153577133  .        T     TAG   117      193303   \n",
       "...          ...   ...        ... ..      ...     ...   ...         ...   \n",
       "1740        1214     1   17393706  .   AAAGAA     AAA  41.8        1214   \n",
       "1741         962     1   14276752  .      GCC      GC  8.73         962   \n",
       "1742         946     1   14106397  .      CCT  CCTCCT  53.9         946   \n",
       "1743         225     1    3697549  .        A  AA,AAA  24.9         225   \n",
       "1744          16     1     760814  .  TTTCTTT     TTT   5.8          16   \n",
       "\n",
       "      InfoNo  \n",
       "0         15  \n",
       "1         15  \n",
       "2         15  \n",
       "3         15  \n",
       "4         15  \n",
       "...      ...  \n",
       "1740      15  \n",
       "1741      15  \n",
       "1742      15  \n",
       "1743      15  \n",
       "1744      15  \n",
       "\n",
       "[1745 rows x 9 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of tag-val pairs per variant \n",
    "# we use group by to group by variant_id in the var_info table\n",
    "# use order by to sort the data\n",
    "\n",
    "# select the max count\n",
    "# SELECT the variants with the max count using a join to get variant data \n",
    "# from the variant table and count for the query that does the count\n",
    "# then use having to only select variant with the count equal to the max\n",
    "\n",
    "select_var_infocount = \"\"\"\n",
    "SELECT *\n",
    "FROM variant v INNER JOIN \n",
    "    (SELECT variant_id, count(*) InfoNo \n",
    "    FROM var_info\n",
    "    GROUP BY variant_id  \n",
    "    HAVING InfoNo IN (SELECT InfoNo FROM \n",
    "                        (SELECT variant_id, count(*) InfoNo \n",
    "                         FROM var_info\n",
    "                         GROUP BY variant_id\n",
    "                         ORDER BY count(*) DESC  \n",
    "                         LIMIT 1))\n",
    "    ORDER BY count(*) DESC\n",
    "    ) s ON s.variant_id = v.variant_id \n",
    "\"\"\"\n",
    "cursor.execute(select_var_infocount)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8633</th>\n",
       "      <td>193545</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8634</th>\n",
       "      <td>193568</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8635</th>\n",
       "      <td>193574</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8636</th>\n",
       "      <td>193584</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8637</th>\n",
       "      <td>193598</td>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8638 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id    tag  value\n",
       "0              3  INDEL      1\n",
       "1              6  INDEL      1\n",
       "2             16  INDEL      1\n",
       "3             30  INDEL      1\n",
       "4             86  INDEL      1\n",
       "...          ...    ...    ...\n",
       "8633      193545  INDEL      1\n",
       "8634      193568  INDEL      1\n",
       "8635      193574  INDEL      1\n",
       "8636      193584  INDEL      1\n",
       "8637      193598  INDEL      1\n",
       "\n",
       "[8638 rows x 3 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select of the DP values for INDEL variants.\n",
    "\n",
    "\n",
    "# select indel variants \n",
    "select_indel = \"\"\"\n",
    "SELECT variant_id, tag, value\n",
    "FROM var_info\n",
    "WHERE tag = 'INDEL';\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "cursor.execute(select_indel)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3305</td>\n",
       "      <td>1</td>\n",
       "      <td>40537941</td>\n",
       "      <td>.</td>\n",
       "      <td>AAA</td>\n",
       "      <td>ACAAA</td>\n",
       "      <td>23.7</td>\n",
       "      <td>3305</td>\n",
       "      <td>DP</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4555</td>\n",
       "      <td>1</td>\n",
       "      <td>59096779</td>\n",
       "      <td>.</td>\n",
       "      <td>GA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>140</td>\n",
       "      <td>4555</td>\n",
       "      <td>DP</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18901</td>\n",
       "      <td>10</td>\n",
       "      <td>5807742</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AA,AAA</td>\n",
       "      <td>53.5</td>\n",
       "      <td>18901</td>\n",
       "      <td>DP</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25814</td>\n",
       "      <td>10</td>\n",
       "      <td>102123890</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "      <td>12.7</td>\n",
       "      <td>25814</td>\n",
       "      <td>DP</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35801</td>\n",
       "      <td>11</td>\n",
       "      <td>130014581</td>\n",
       "      <td>.</td>\n",
       "      <td>TTT</td>\n",
       "      <td>TT</td>\n",
       "      <td>90.5</td>\n",
       "      <td>35801</td>\n",
       "      <td>DP</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>38296</td>\n",
       "      <td>12</td>\n",
       "      <td>19566184</td>\n",
       "      <td>.</td>\n",
       "      <td>GTTT</td>\n",
       "      <td>GTT</td>\n",
       "      <td>88.5</td>\n",
       "      <td>38296</td>\n",
       "      <td>DP</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>40640</td>\n",
       "      <td>12</td>\n",
       "      <td>53435294</td>\n",
       "      <td>.</td>\n",
       "      <td>TTTTTTT</td>\n",
       "      <td>TTTTTTTTTT</td>\n",
       "      <td>150</td>\n",
       "      <td>40640</td>\n",
       "      <td>DP</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>53399</td>\n",
       "      <td>14</td>\n",
       "      <td>50329569</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>GAAT,GCT</td>\n",
       "      <td>44.5</td>\n",
       "      <td>53399</td>\n",
       "      <td>DP</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>65374</td>\n",
       "      <td>16</td>\n",
       "      <td>33964064</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TC</td>\n",
       "      <td>202</td>\n",
       "      <td>65374</td>\n",
       "      <td>DP</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>81965</td>\n",
       "      <td>19</td>\n",
       "      <td>13054921</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "      <td>20.5</td>\n",
       "      <td>81965</td>\n",
       "      <td>DP</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>103233</td>\n",
       "      <td>20</td>\n",
       "      <td>26189984</td>\n",
       "      <td>.</td>\n",
       "      <td>CAA</td>\n",
       "      <td>CAAA</td>\n",
       "      <td>140</td>\n",
       "      <td>103233</td>\n",
       "      <td>DP</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>118041</td>\n",
       "      <td>3</td>\n",
       "      <td>96336847</td>\n",
       "      <td>.</td>\n",
       "      <td>GGG</td>\n",
       "      <td>GCCCGCGG</td>\n",
       "      <td>4.94</td>\n",
       "      <td>118041</td>\n",
       "      <td>DP</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>137100</td>\n",
       "      <td>5</td>\n",
       "      <td>33162469</td>\n",
       "      <td>.</td>\n",
       "      <td>TGGGGG</td>\n",
       "      <td>TGGGG</td>\n",
       "      <td>211</td>\n",
       "      <td>137100</td>\n",
       "      <td>DP</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>142650</td>\n",
       "      <td>5</td>\n",
       "      <td>151043160</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "      <td>71.5</td>\n",
       "      <td>142650</td>\n",
       "      <td>DP</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>147772</td>\n",
       "      <td>6</td>\n",
       "      <td>30558477</td>\n",
       "      <td>.</td>\n",
       "      <td>GA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>182</td>\n",
       "      <td>147772</td>\n",
       "      <td>DP</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>147968</td>\n",
       "      <td>6</td>\n",
       "      <td>31236716</td>\n",
       "      <td>.</td>\n",
       "      <td>CCCC</td>\n",
       "      <td>CCC</td>\n",
       "      <td>138</td>\n",
       "      <td>147968</td>\n",
       "      <td>DP</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>149196</td>\n",
       "      <td>6</td>\n",
       "      <td>35801733</td>\n",
       "      <td>.</td>\n",
       "      <td>ATTACTTACT</td>\n",
       "      <td>ATTACT</td>\n",
       "      <td>140</td>\n",
       "      <td>149196</td>\n",
       "      <td>DP</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>149197</td>\n",
       "      <td>6</td>\n",
       "      <td>35801923</td>\n",
       "      <td>.</td>\n",
       "      <td>TAACGAAC</td>\n",
       "      <td>TAAC</td>\n",
       "      <td>46.5</td>\n",
       "      <td>149197</td>\n",
       "      <td>DP</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>154397</td>\n",
       "      <td>6</td>\n",
       "      <td>114181918</td>\n",
       "      <td>.</td>\n",
       "      <td>TTTTTTTTT</td>\n",
       "      <td>TTTTTTT</td>\n",
       "      <td>54.5</td>\n",
       "      <td>154397</td>\n",
       "      <td>DP</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>154399</td>\n",
       "      <td>6</td>\n",
       "      <td>114182385</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>61.5</td>\n",
       "      <td>154399</td>\n",
       "      <td>DP</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>175602</td>\n",
       "      <td>8</td>\n",
       "      <td>124513622</td>\n",
       "      <td>.</td>\n",
       "      <td>AACAAAA</td>\n",
       "      <td>AACAAAAAAAAACAAAA</td>\n",
       "      <td>18.5</td>\n",
       "      <td>175602</td>\n",
       "      <td>DP</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>181938</td>\n",
       "      <td>9</td>\n",
       "      <td>35658015</td>\n",
       "      <td>.</td>\n",
       "      <td>CAC</td>\n",
       "      <td>CACAC</td>\n",
       "      <td>19.5</td>\n",
       "      <td>181938</td>\n",
       "      <td>DP</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id CHROM        POS ID         REF                ALT  QUAL  \\\n",
       "0         3305     1   40537941  .         AAA              ACAAA  23.7   \n",
       "1         4555     1   59096779  .          GA                GAA   140   \n",
       "2        18901    10    5807742  .           A             AA,AAA  53.5   \n",
       "3        25814    10  102123890  .           T                 TT  12.7   \n",
       "4        35801    11  130014581  .         TTT                 TT  90.5   \n",
       "5        38296    12   19566184  .        GTTT                GTT  88.5   \n",
       "6        40640    12   53435294  .     TTTTTTT         TTTTTTTTTT   150   \n",
       "7        53399    14   50329569  .          GT           GAAT,GCT  44.5   \n",
       "8        65374    16   33964064  .           T                 TC   202   \n",
       "9        81965    19   13054921  .           T                 TT  20.5   \n",
       "10      103233    20   26189984  .         CAA               CAAA   140   \n",
       "11      118041     3   96336847  .         GGG           GCCCGCGG  4.94   \n",
       "12      137100     5   33162469  .      TGGGGG              TGGGG   211   \n",
       "13      142650     5  151043160  .           T                 TT  71.5   \n",
       "14      147772     6   30558477  .          GA                GAA   182   \n",
       "15      147968     6   31236716  .        CCCC                CCC   138   \n",
       "16      149196     6   35801733  .  ATTACTTACT             ATTACT   140   \n",
       "17      149197     6   35801923  .    TAACGAAC               TAAC  46.5   \n",
       "18      154397     6  114181918  .   TTTTTTTTT            TTTTTTT  54.5   \n",
       "19      154399     6  114182385  .           A                 AA  61.5   \n",
       "20      175602     8  124513622  .     AACAAAA  AACAAAAAAAAACAAAA  18.5   \n",
       "21      181938     9   35658015  .         CAC              CACAC  19.5   \n",
       "\n",
       "    variant_id tag  value  \n",
       "0         3305  DP    158  \n",
       "1         4555  DP    163  \n",
       "2        18901  DP    180  \n",
       "3        25814  DP    153  \n",
       "4        35801  DP    159  \n",
       "5        38296  DP    190  \n",
       "6        40640  DP    162  \n",
       "7        53399  DP    204  \n",
       "8        65374  DP    179  \n",
       "9        81965  DP    151  \n",
       "10      103233  DP    155  \n",
       "11      118041  DP    205  \n",
       "12      137100  DP    214  \n",
       "13      142650  DP    227  \n",
       "14      147772  DP    221  \n",
       "15      147968  DP    159  \n",
       "16      149196  DP    193  \n",
       "17      149197  DP    220  \n",
       "18      154397  DP    153  \n",
       "19      154399  DP    192  \n",
       "20      175602  DP    210  \n",
       "21      181938  DP    242  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select of INDEL variants with DP values > 150.\n",
    "\n",
    "\n",
    "# select variant data from variant table and \n",
    "# and take only the variants that are indels\n",
    "# inner join with the var_info table to get the dp values  \n",
    "\n",
    "\n",
    "select_indel = \"\"\"\n",
    "SELECT *\n",
    "FROM variant \n",
    "    INNER JOIN var_info ON variant.variant_id = var_info.variant_id\n",
    "WHERE\n",
    "    tag = 'DP' AND\n",
    "    value > 150 AND\n",
    "    variant.variant_id IN \n",
    "        (SELECT variant_id\n",
    "        FROM var_info\n",
    "        WHERE tag = 'INDEL');\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "cursor.execute(select_indel)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, CHROM, POS, ID, REF, ALT, QUAL]\n",
       "Index: []"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select of variants with rsid\n",
    "\n",
    "select_rs = \"\"\"\n",
    "SELECT *\n",
    "FROM variant \n",
    "WHERE ID LIKE 'rs%';\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "cursor.execute(select_rs)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variant_id, CHROM, POS, ID, REF, ALT, QUAL]\n",
       "Index: []"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# A select of variants with id not dot\n",
    "\n",
    "select_rs = \"\"\"\n",
    "SELECT *\n",
    "FROM variant \n",
    "WHERE ID <> \".\";\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "cursor.execute(select_rs)\n",
    "\n",
    "data = cursor.fetchall()\n",
    "header = [ct[0] for ct in cursor.description]\n",
    "df_res = pd.DataFrame(data, columns = header)\n",
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we are done working with this database\n",
    "# close cursor and connection\n",
    "\n",
    "cursor.close()\n",
    "connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
