{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Study session 9 - simply bioinformatics\n",
    "#### Use data and control structures, pandas, dataframes and explaore new libraries\n",
    "#### Explore object  attributes and methods\n",
    "#### BIOINF 575 - Fall 2021\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "____\n",
    "\n",
    "#### Variant data exploration   \n",
    "We have variant data for a sample in the file `variant_data_file.vcf`.   \n",
    "The vcf data file contains information about differences found in the genome of a specific sample when compared to the reference.    \n",
    "The file contains approximatelly 1000 differences found in the Y chromosome.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "..................................................................................................\n",
    "\n",
    "<b><font color = \"red\">Exercise</font> <br></b>\n",
    "\n",
    "<b>Total depth distribution</b>\n",
    "\n",
    "We want to have a look at the distribution of the total depth in our data (DP key in the INFO column).\n",
    "- Read the data into a pandas data frame\n",
    "    - look at the comment and names arguments for the read_csv funtion in the pandas library, you will need to add the header by hand\n",
    "- Write a function that selects the total depth from the INFO part of a variant (row in our dataframe)\n",
    "- Apply the function for each row of the dataframe which will result in a pandas series with the total depth for each variant \n",
    "- Plot the histogram of total depths (pandas series previously computed)\n",
    "- Plot the histogram of the total depths that are less than 170"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import necessary libraries\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>2728456</td>\n",
       "      <td>rs2058276</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>32</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AN=2;DB;DP=182;H2;NS=65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>2734240</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AN=2;DP=196;NS=63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>2743242</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AN=2;DP=275;NS=66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>2746727</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AN=2;DP=179;NS=64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>2777970</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>67</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AN=2;DP=225;NS=67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954</th>\n",
       "      <td>Y</td>\n",
       "      <td>27108406</td>\n",
       "      <td>rs2049642</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>30</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AN=2;DB;DP=163;NS=61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>Y</td>\n",
       "      <td>27141305</td>\n",
       "      <td>rs9786602</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>29</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AN=2;DB;DP=163;NS=63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>956</th>\n",
       "      <td>Y</td>\n",
       "      <td>27142489</td>\n",
       "      <td>rs1358368</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>37</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=2;AN=2;DB;DP=195;H2;NS=61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>Y</td>\n",
       "      <td>27182557</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>AC=1;AN=2;DP=168;NS=62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>958</th>\n",
       "      <td>Y</td>\n",
       "      <td>27228657</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>6</td>\n",
       "      <td>NUYR</td>\n",
       "      <td>AC=1;AN=2;DP=317;NS=65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>959 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    CHROM       POS         ID REF ALT  QUAL FILTER  \\\n",
       "0       Y   2728456  rs2058276   T   C    32      .   \n",
       "1       Y   2734240          .   G   A    31      .   \n",
       "2       Y   2743242          .   C   T    25      .   \n",
       "3       Y   2746727          .   A   G    34      .   \n",
       "4       Y   2777970          .   T   A    67      .   \n",
       "..    ...       ...        ...  ..  ..   ...    ...   \n",
       "954     Y  27108406  rs2049642   G   A    30      .   \n",
       "955     Y  27141305  rs9786602   C   A    29      .   \n",
       "956     Y  27142489  rs1358368   G   C    37      .   \n",
       "957     Y  27182557          .   G   A    22      .   \n",
       "958     Y  27228657          .   G   T     6   NUYR   \n",
       "\n",
       "                             INFO  \n",
       "0    AC=2;AN=2;DB;DP=182;H2;NS=65  \n",
       "1          AC=1;AN=2;DP=196;NS=63  \n",
       "2          AC=1;AN=2;DP=275;NS=66  \n",
       "3          AC=2;AN=2;DP=179;NS=64  \n",
       "4          AC=1;AN=2;DP=225;NS=67  \n",
       "..                            ...  \n",
       "954     AC=1;AN=2;DB;DP=163;NS=61  \n",
       "955     AC=2;AN=2;DB;DP=163;NS=63  \n",
       "956  AC=2;AN=2;DB;DP=195;H2;NS=61  \n",
       "957        AC=1;AN=2;DP=168;NS=62  \n",
       "958        AC=1;AN=2;DP=317;NS=65  \n",
       "\n",
       "[959 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read the data from the vcf file  and add header (column labels)\n",
    "# pd.read_csv function has a names argument that allows us to set the column labels\n",
    "\n",
    "columns = (\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\")\n",
    "vcf_data = pd.read_csv(\"variant_data_file.vcf\", comment = \"#\", sep = \"\\t\", names = columns)\n",
    "vcf_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write a function that return the value for the DP key\n",
    "# from the INFO part of a variant data row/line \n",
    "# The info key=value pairs are separated by \";\"\n",
    "# There might be flag values, that do not have a value\n",
    "# In that case there is no = in the pair data e.g.: ;DB;\n",
    "\n",
    "def get_dp(info):\n",
    "    '''\n",
    "    Function that returns the total depth value for a variant from the variant information\n",
    "    Args:\n",
    "    info (str): the INFO column/element for a variant/data line from a vcf file\n",
    "    Returns:\n",
    "    int: the total depth value. for the variant, the value for DP key\n",
    "    '''\n",
    "    dp = np.nan\n",
    "    for pair in info.split(\";\"):\n",
    "        if \"=\" in pair:\n",
    "            key, val = pair.split(\"=\")\n",
    "            if key == \"DP\":\n",
    "                return int(val)\n",
    "    return dp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "182"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the function\n",
    "get_dp(\"AC=2;AN=2;DB;DP=182;H2;NS=65\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      182\n",
       "1      196\n",
       "2      275\n",
       "3      179\n",
       "4      225\n",
       "      ... \n",
       "954    163\n",
       "955    163\n",
       "956    195\n",
       "957    168\n",
       "958    317\n",
       "Name: INFO, Length: 959, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Use the apply method for a pandas Series (dataframe column)\n",
    "# to call the funtion for each element in the Series/column\n",
    "# and collect the result in the resulting Series \n",
    "# the resulting series has the same name as the one the function was applied on\n",
    "\n",
    "dp_data = vcf_data.INFO.apply(get_dp)\n",
    "dp_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAD4CAYAAADrRI2NAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAARFUlEQVR4nO3df6zddX3H8efLwpBwnZSANx0lK9m6ZUAzlBtmQrKcihtEjGXLSGrQlISl+wMTzTCz+I/6R7NuCWZ/IMs6MWuGetOghgbCFla9MSYypIqWgoxOOlYgbVRArzEkxff+ON+OA97b++uce9v7eT6Sm/P9fs7nc76f8+4nr/O933vOaaoKSVI73rLSE5AkLS+DX5IaY/BLUmMMfklqjMEvSY05a6UnAHDhhRfWhg0b+MUvfsF555230tNZcdbhddaizzr0WYe+k3U4cODAj6vqooWOPy2Cf8OGDTz22GNMTU3R6/VWejorzjq8zlr0WYc+69B3sg5J/mcx4+d1qSfJkSQHkzye5LGu7YIkDyd5prtdO9D/jiSHkzyd5LrFTEySNBoLuca/uaqurKqJbn8HsL+qNgL7u32SXAZsBS4HrgfuTrJmiHOWJC3BUv64uwXY023vAW4caJ+sqler6lngMHD1Eo4jSRqizOcrG5I8C7wEFPBPVbU7yctVdf5An5eqam2Su4BHqurerv0e4KGquu9Nj7kd2A4wPj5+1eTkJNPT04yNjQ3ruZ2xrMPrrEWfdeizDn0n67B58+YDA1dh5m2+f9y9pqpeSPIO4OEkPzxF38zQ9muvLlW1G9gNMDExUb1ezz/cdKzD66xFn3Xosw59S63DvC71VNUL3e1x4Gv0L90cS7IOoLs93nU/ClwyMHw98MKiZyhJGqo5gz/JeUnednIb+FPgCWAfsK3rtg24v9veB2xNck6SS4GNwKPDnrgkaXHmc6lnHPhakpP9v1RV/5bkO8DeJLcCzwE3AVTVoSR7gSeBE8BtVfXaSGYvSVqwOYO/qn4E/OEM7T8Brp1lzE5g55JnJ0kautPik7tLtWHHgyty3CO7bliR40rSUvglbZLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1Jh5B3+SNUm+l+SBbv+CJA8neaa7XTvQ944kh5M8neS6UUxckrQ4Cznj/yjw1MD+DmB/VW0E9nf7JLkM2ApcDlwP3J1kzXCmK0laqnkFf5L1wA3A5weatwB7uu09wI0D7ZNV9WpVPQscBq4eymwlSUuWqpq7U3If8LfA24CPV9X7k7xcVecP9HmpqtYmuQt4pKru7drvAR6qqvve9Jjbge0A4+PjV01OTjI9Pc3Y2NiCn8TB519Z8Jhh2HTx20fyuIutw2pkLfqsQ5916DtZh82bNx+oqomFjj9rrg5J3g8cr6oDSXrzeMzM0PZrry5VtRvYDTAxMVG9Xo+pqSl6vfkc4o1u2fHggscMw5GbeyN53MXWYTWyFn3Woc869C21DnMGP3AN8IEk7wPeCvxmknuBY0nWVdWLSdYBx7v+R4FLBsavB15Y9AwlSUM15zX+qrqjqtZX1Qb6f7T9elV9CNgHbOu6bQPu77b3AVuTnJPkUmAj8OjQZy5JWpT5nPHPZhewN8mtwHPATQBVdSjJXuBJ4ARwW1W9tuSZSpKGYkHBX1VTwFS3/RPg2ln67QR2LnFukqQR8JO7ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUmDmDP8lbkzya5PtJDiX5TNd+QZKHkzzT3a4dGHNHksNJnk5y3SifgCRpYc6aR59XgfdU1XSSs4FvJXkI+HNgf1XtSrID2AF8IsllwFbgcuC3gP9I8ntV9dqInsOK2bDjwZE87u2bTnDLKR77yK4bRnJcSW2Y84y/+qa73bO7nwK2AHu69j3Ajd32FmCyql6tqmeBw8DVw5y0JGnxUlVzd0rWAAeA3wU+V1WfSPJyVZ0/0Oelqlqb5C7gkaq6t2u/B3ioqu5702NuB7YDjI+PXzU5Ocn09DRjY2MLfhIHn39lwWNOZ+PnwrFfzn7/povfvnyTWWGLXROrjXXosw59J+uwefPmA1U1sdDx87nUQ3eZ5sok5wNfS3LFKbpnpoeY4TF3A7sBJiYmqtfrMTU1Ra/Xm8+U3uBUl0XORLdvOsGdB2f/pzlyc2/5JrPCFrsmVhvr0Gcd+pZahwW9q6eqXgamgOuBY0nWAXS3x7tuR4FLBoatB15Y9AwlSUM1n3f1XNSd6ZPkXOC9wA+BfcC2rts24P5uex+wNck5SS4FNgKPDnnekqRFms+lnnXAnu46/1uAvVX1QJJvA3uT3Ao8B9wEUFWHkuwFngROALetxnf0SNKZas7gr6ofAO+cof0nwLWzjNkJ7Fzy7CRJQ+cndyWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1Zs7gT3JJkm8keSrJoSQf7dovSPJwkme627UDY+5IcjjJ00muG+UTkCQtzHzO+E8At1fVHwDvBm5LchmwA9hfVRuB/d0+3X1bgcuB64G7k6wZxeQlSQs3Z/BX1YtV9d1u++fAU8DFwBZgT9dtD3Bjt70FmKyqV6vqWeAwcPWQ5y1JWqRU1fw7JxuAbwJXAM9V1fkD971UVWuT3AU8UlX3du33AA9V1X1veqztwHaA8fHxqyYnJ5menmZsbGzBT+Lg868seMzpbPxcOPbL2e/fdPHbl28yK2yxa2K1sQ591qHvZB02b958oKomFjr+rPl2TDIGfAX4WFX9LMmsXWdo+7VXl6raDewGmJiYqF6vx9TUFL1eb75T+n+37HhwwWNOZ7dvOsGdB2f/pzlyc2/5JrPCFrsmVhvr0Gcd+pZah3m9qyfJ2fRD/4tV9dWu+ViSdd3964DjXftR4JKB4euBFxY9Q0nSUM3nXT0B7gGeqqrPDty1D9jWbW8D7h9o35rknCSXAhuBR4c3ZUnSUsznUs81wIeBg0ke79o+CewC9ia5FXgOuAmgqg4l2Qs8Sf8dQbdV1WvDnrgkaXHmDP6q+hYzX7cHuHaWMTuBnUuYlyRpRPzkriQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWrMWSs9AS3chh0Prtixj+y6YcWOLWk4POOXpMYY/JLUmDmDP8kXkhxP8sRA2wVJHk7yTHe7duC+O5IcTvJ0kutGNXFJ0uLM54z/X4Dr39S2A9hfVRuB/d0+SS4DtgKXd2PuTrJmaLOVJC3ZnMFfVd8Efvqm5i3Anm57D3DjQPtkVb1aVc8Ch4GrhzNVSdIwpKrm7pRsAB6oqiu6/Zer6vyB+1+qqrVJ7gIeqap7u/Z7gIeq6r4ZHnM7sB1gfHz8qsnJSaanpxkbG1vwkzj4/CsLHnM6Gz8Xjv1ypWcxs00Xv31Zj7fYNbHaWIc+69B3sg6bN28+UFUTCx0/7LdzZoa2GV9Zqmo3sBtgYmKier0eU1NT9Hq9BR/0lhV8e+Mo3L7pBHcePD3faXvk5t6yHm+xa2K1sQ591qFvqXVY7Lt6jiVZB9DdHu/ajwKXDPRbD7yw6NlJkoZuscG/D9jWbW8D7h9o35rknCSXAhuBR5c2RUnSMM15PSHJl4EecGGSo8CngF3A3iS3As8BNwFU1aEke4EngRPAbVX12ojmLklahDmDv6o+OMtd187SfyewcymTkiSNjp/claTGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNeaslZ6Aziwbdjy4rMe7fdMJbtnxIEd23bCsx5VWM8/4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY3xA1w6Iyz3B8cG+eExrTYjO+NPcn2Sp5McTrJjVMeRJC3MSM74k6wBPgf8CXAU+E6SfVX15CiOJ61Gg7/lnPzqiuXibzmr26gu9VwNHK6qHwEkmQS2AAa/zjgreZlppZyuz3m5XwCXw0q8yKaqhv+gyV8A11fVX3b7Hwb+qKo+MtBnO7C92/194GngQuDHQ5/Qmcc6vM5a9FmHPuvQd7IOv11VFy108KjO+DND2xteYapqN7D7DYOSx6pqYkRzOmNYh9dZiz7r0Gcd+pZah1H9cfcocMnA/nrghREdS5K0AKMK/u8AG5NcmuQ3gK3AvhEdS5K0ACO51FNVJ5J8BPh3YA3whao6NI+hu+fu0gTr8Dpr0Wcd+qxD35LqMJI/7kqSTl9+ZYMkNcbgl6TGLGvwJ/lCkuNJnhho+3SS55M83v28b+C+O7qvfHg6yXXLOddRSnJJkm8keSrJoSQf7dovSPJwkme627UDY1ZdLU5Rh6bWRJK3Jnk0yfe7Onyma29tPcxWh6bWw0lJ1iT5XpIHuv3hrYeqWrYf4I+BdwFPDLR9Gvj4DH0vA74PnANcCvw3sGY55zvCOqwD3tVtvw34r+75/j2wo2vfAfzdaq7FKerQ1Jqg/7mXsW77bOA/gXc3uB5mq0NT62Hg+f018CXggW5/aOthWc/4q+qbwE/n2X0LMFlVr1bVs8Bh+l8Fccarqher6rvd9s+Bp4CL6T/nPV23PcCN3faqrMUp6jCb1VqHqqrpbvfs7qdobz3MVofZrMo6ACRZD9wAfH6geWjr4XS5xv+RJD/oLgWd/PXlYuB/B/oc5dShcEZKsgF4J/2zm/GqehH6oQi8o+u26mvxpjpAY2ui+7X+ceA48HBVNbkeZqkDNLYegH8A/gb41UDb0NbD6RD8/wj8DnAl8CJwZ9c+59c+nOmSjAFfAT5WVT87VdcZ2lZNLWaoQ3Nroqpeq6or6X/K/eokV5yie2t1aGo9JHk/cLyqDsx3yAxtp6zDigd/VR3r/rF/Bfwzr/+Ksqq/9iHJ2fTD7otV9dWu+ViSdd396+if9cAqrsVMdWh1TQBU1cvAFHA9Da6Hkwbr0OB6uAb4QJIjwCTwniT3MsT1sOLBf/KJdP4MOPmOn33A1iTnJLkU2Ag8utzzG4UkAe4Bnqqqzw7ctQ/Y1m1vA+4faF91tZitDq2tiSQXJTm/2z4XeC/wQ9pbDzPWobX1UFV3VNX6qtpA/+tuvl5VH2KI62FZ/+vFJF8GesCFSY4CnwJ6Sa6k/6vJEeCvAKrqUJK99L/D/wRwW1W9tpzzHaFrgA8DB7vrmQCfBHYBe5PcCjwH3ASruhaz1eGDja2JdcCe9P8Do7cAe6vqgSTfpq31MFsd/rWx9TCboeWDX9kgSY1Z8Us9kqTlZfBLUmMMfklqjMEvSY0x+CWpMQa/JDXG4JekxvwfhPuHvwfIt2wAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot the histogram for a pandas Series using the hist method\n",
    "# we see a very steep decrese in frequency as the total depth values increase\n",
    "\n",
    "dp_data.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEWCAYAAABhffzLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAa80lEQVR4nO3de5hcVZnv8e+PcAsJJCDQTxLBBg+gSCBzaJkRATtcFEc9iAOHm1wEJ+qoo544CIznDCpqRokwgp4ZOCKMIPHCXVEGGUIcdMSEAZIQGSCEXIgJkBBIuAbe88dePakuqrsr1b2rumv9Ps9TT9fee+293l61661Vq3atUkRgZmb52KLVAZiZWXM58ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+G3ISLpS0gV1lu2UFJK2LCmWkPTfyjh2K0haIunIPrYdKumhZsdkI5cTf0b6Sx5DUb5VJM2W9NEW1DtV0p2S1klaUrVtd0nrq24haXpFmZMlPS5pg6QbJe3USBwR8euI2KeOeM+XdHUjdVh7ceI364ekcZK26WPzBuAK4G+qN0TE0ogY23MDJgOvAdel474N+CfgVKADeB74bgn/QlOU9c7NyuHEnwlJPwB2B25Jvc+z0/r/IWmhpGdSz/mtA5T/iaQ/pl7unJTA6ql/lKQLJT0laTHwvqrt4yR9T9JKSSskXSBpVNp2hqS7JV2S6v2DpCPStq8ChwKXpjgvrTjskZIelrRW0nckqc5Yt5B0pKRrgGXAzrXKRcQ9EfEDYHEdhz0NmBMRS9LyKcAtETEnItYD/xv4kKTt+znGFEkPpDb4kaRtU7zdkpZXxP+F1IbPSXpI0hGSjgbOA05I7XR/KjtR0s2S1kh6RNJfVhxntKSrUvstknR2VT1LUl0PABskbSnpHEmPproflHRsRfmex/GidL4tlnRwWr9M0mpJp9fRljZYEeFbJjdgCXBkxfLeFL3Wo4CtgLOBR4Cta5VP684Etge2AS4G7qvYdiVwQR91fxz4A7AbsBNwJxDAlmn7jRQ94DHArsA9wMfStjOAjcDnUpwnAOuAndL22cBHq+oL4GfAeIoXsCeBowdonz2BLwOPAw8A04GOOtr1SGDJAGUeBc6oWL4J+EJVmfXAgf08dvcAE1P7LQI+nrZ1A8vT/X0oXqwmpuVO4M3p/vnA1VXHvYvinca2wJTUTkekbTPS9h2BN6Y2WV4V033pMR2d1h2fYtwiPU4bgAlVj+NHgFHABcBS4DvpfHo38BwwttXPlXa/uceftxOAn0fE7RHxCnAhMBo4uK8dIuKKiHguIl6iSCQHSBpXR13/E7g4IpZFxBrg6z0bJHUA7wU+GxEbImI1cBFwYsX+q9P+r0TEj4CHqHrXUMOMiHgmIpZSvNBMqVVI0gGSZgP/TvFCcWxE7B8RMyNiVR3/W78kHUoxnPPTitVjKV68Kq2jeFHty7cj4onUfrdQ+/95lSKJ7itpq4hYEhGP9hHXbsAhFC9AL0bEfcD/oxh+guIx+1pErI2I5cC3+4hpWUS8ABARP0kxvpYep4eBgyrKPxYR34+IV4EfUbxofDkiXoqIfwFeBtrmQ/nhyok/bxMpercARMRrFL3FSbUKp+GaGemt/LMUPT7oYyikRl3LKpYfr7j/Joqe/Mo0BPAMRe9/14oyKyJ1Gyv2nzhAnX+suP88RbKtZTzwFop3O/env0PpdOC6KIZ0eqwHdqgqtwNFj7cvA/4/EfEI8FmKF+XVkmZJ6qudJgJrIqKyzsfZ9PhXP2aV92uuk3SapPsqHsf96H1+VL6Q9rxYVK/r63GyIeLEn5fqqVifoEi6AKQx8N2AFX2UPxk4hmJoYxzFMAJAPWPnK9Oxe+xecX8Z8BKwc0SMT7cdIqLy84NJVWP0u6f4a8W5WSLiLoqhjK9TvItYKulaSUf3fM7QKEmjKYY/rqratBA4oKLcnhQ99f8cTH0AEfHDiDiE4rEN4O97NlUVfQLYqepzhd3Z9PivpGiXHpWP339V13NH0puAy4FPAW+IiPHAAuo7P6yJnPjzsopiHLvHj4H3pQ//tqIY034J+E0f5bdP258GtgO+thl1/xj4a0lvlLQjcE7PhohYCfwLMFPSDunD1TdLelfF/rum/beSdDzwVuDWPuLcbBGxMSJuiYgPUQw1/DvFC8EySbvW2ifFuS3FuxVJ2lbS1lXFjgWeoRhqqnQN8AEV1+CPofhs4fqq3vdmk7SPpMNVXIn0IkUP+tW0eRXQKWmL9D8vo3isv55i3x84K8UGxWN2rqQdJU2iSOj9GUPxQvBkiuUjFD1+G2ac+PPydeCL6W345yPiIeDDwCXAU8AHgA9ExMu1ygP/TDEUsAJ4kCI51uty4DaKoZR7geurtp8GbJ2Ou5ZiPHxCxfbfAXulOL8KHBcRT6dt/wAcl64+qTUOvVki4qmI+IeI+BOKzx6e76PoYRSJ9VaKnvILFC9glU4H/rlqmIqIWEjxgfc1FJ9fbA/81WBjp3jXMIOinf5I8YJ5Xtr2k/T3aUn3pvsnUbxzewK4Afi7iLg9bfsysBx4DPgVxWPyUl8VR8SDwEzgtxQvMpOBu4fgf7Ihpqrz0WzYkXQGxVU7h7Q6lpxJ+gRwYkS8a8DCNqy5x29mNUmaIOmdaUhrH4qhwBtaHZcNnr9tZ2Z92Zri6qo9KD6nmMUI/naxbeKhHjOzzHiox8wsMyNiqGfnnXeOzs7OhvbdsGEDY8aMGdqARjC3xyZui97cHr21Q3vMmzfvqYjYpXr9iEj8nZ2dzJ07t6F9Z8+eTXd399AGNIK5PTZxW/Tm9uitHdpD0uO11nuox8wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDMj4pu7NjJ0nvPzltW9ZMZAv7tu7aCZ59j0yRs5I9XXbueXe/xmZplx4jczy4wTv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMeFpmawuNTNdbOe1uo9ptul7Lg3v8ZmaZceI3M8uME7+ZWWZKS/ySdpN0p6RFkhZK+kxaf76kFZLuS7c/LysGMzN7vTI/3N0ITI+IeyVtD8yTdHvadlFEXFhi3WZm1ofSEn9ErARWpvvPSVoETCqrPjMzq48iovxKpE5gDrAf8L+AM4BngbkU7wrW1thnGjANoKOj48BZs2Y1VPf69esZO3ZsQ/u2ozLbY/6KdaUctywdo2HVC4M7xuRJ44YmmGFgJDxXmnmODcX5MRQGc45NnTp1XkR0Va8vPfFLGgvcBXw1Iq6X1AE8BQTwFWBCRJzZ3zG6urpi7ty5DdU/e/Zsuru7G9q3HZXZHo1cS99K0ydvZOb8wb3pbafr+EfCc6WZ59hQnB9DYTDnmKSaib/Uq3okbQVcB1wTEdcDRMSqiHg1Il4DLgcOKjMGMzPrrcyregR8D1gUEd+qWD+hotixwIKyYjAzs9cr833MO4FTgfmS7kvrzgNOkjSFYqhnCfCxEmMwM7MqZV7V82+Aamy6taw6zcxsYP7mrplZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWZaP/VcG2vVbJXtNGPkcNfKGUn9OFuj3OM3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWWm7S/nnL9iHWeMsB8BNzMrk3v8ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMtP3lnDnqb8bI6ZM3+vJWG5RWzkhqQ8M9fjOzzDjxm5llxonfzCwzTvxmZpkpLfFL2k3SnZIWSVoo6TNp/U6Sbpf0cPq7Y1kxmJnZ65XZ498ITI+ItwJ/BnxS0r7AOcAdEbEXcEdaNjOzJikt8UfEyoi4N91/DlgETAKOAa5Kxa4CPlhWDGZm9nqKiPIrkTqBOcB+wNKIGF+xbW1EvG64R9I0YBpAR0fHgbNmzWqo7tVr1rHqhYZ2bUsdo3F7JCO9LSZPGjekx1u/fj1jx44dsNz8FeuGtN7haricH4N5nKdOnTovIrqq15f+BS5JY4HrgM9GxLOS6tovIi4DLgPo6uqK7u7uhuq/5JqbmDnf31PrMX3yRrdHMtLbYskp3UN6vNmzZ1PP8yyXLwAOl/NjqB9nKPmqHklbUST9ayLi+rR6laQJafsEYHWZMZiZWW9lXtUj4HvAooj4VsWmm4HT0/3TgZvKisHMzF6vzPcx7wROBeZLui+tOw+YAfxY0lnAUuD4EmMwM7MqpSX+iPg3oK8B/SPKqtfMzPrnb+6amWWm9R9Zm1lDhnp6ZE/ZnQ/3+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmXHiNzPLjBO/mVlmnPjNzDIzYOKXNFfSJ/3buGZm7aGeHv+JwETg95JmSXqP6v01FTMzG3YGTPwR8UhE/C2wN/BD4ApgqaQvSdqp7ADNzGxo1TXGL2l/YCbwTYpf1DoOeBb41/JCMzOzMgw4O6ekecAzFL+mdU5EvJQ2/U7SO0uMzczMSlDPtMzHR8TiWhsi4kNDHI+ZmZWsnqGej0oa37MgaUdJF5QXkpmZlamexP/eiHimZyEi1gJ/XlpEZmZWqnoS/yhJ2/QsSBoNbNNPeTMzG8bqGeO/GrhD0veBAM4Erio1KjMzK82AiT8iviFpPnAEIOArEXFb6ZGZmVkp6vqx9Yj4BfCLkmMxM7MmqGeung9JeljSOknPSnpO0rPNCM7MzIZePT3+bwAfiIhFZQdjZmblq+eqnlVO+mZm7aOeHv9cST8CbgR6pmsgIq4vKygzMytPPYl/B+B54N0V6wJw4jczG4HquZzzI40cWNIVwPuB1RGxX1p3PvCXwJOp2HkRcWsjxzczs8bUc1XP3pLukLQgLe8v6Yt1HPtK4Oga6y+KiCnp5qRvZtZk9Xy4ezlwLvAKQEQ8QPGrXP2KiDnAmkFFZ2ZmQ66eMf7tIuKeql9b3DiIOj8l6TRgLjA9Tfr2OpKmAdMAOjo6mD17dkOVdYyG6ZMHE257cXts4rboze3R23Bpj0ZzX3/qSfxPSXozxQe6SDoOWNlgff8X+Eo61lcoftXrzFoFI+Iy4DKArq6u6O7ubqjCS665iZnz6/qCchamT97o9kjcFr25PXobLu2x5JTuIT9mPf/VJykS8FskrQAeAz7cSGURsarnvqTLgZ81chwzM2tcPVf1LAaOlDQG2CIinmu0MkkTIqLn3cKxwIJGj2VmZo2p5zd3/0/VMgAR8eUB9rsW6AZ2lrQc+DugW9IUiqGeJcDHGojZzMwGoZ6hng0V97eluDZ/wCkcIuKkGqu/V2dcZmZWknqGemZWLku6ELi5tIjMzKxU9VzHX207YM+hDsTMzJqjnjH++aRLOYFRwC5Av+P7ZmY2fNUzxv/+ivsbKaZpbv23GszMrCH1JP7qyzd3qPwWb0R4WgYzsxGknsR/L7AbsJbix9bHA0vTtsDj/WZmI0o9H+7+kuKnF3eOiDdQDP1cHxF7RISTvpnZCFNP4n975fTJEfEL4F3lhWRmZmWqd5K2LwJXUwztfBh4utSozMysNPX0+E+iuITzhnTbJa0zM7MRqJ5v7q4BPiNpbESsb0JMZmZWonp+evFgSQ8CD6blAyR9t/TIzMysFPUM9VwEvIc0rh8R9wOHlRmUmZmVp665eiJiWdWqV0uIxczMmqCeq3qWSToYCElbA39NHdMym5nZ8FRPj//jFD+/OAlYDkxJy2ZmNgL12+OXNAq4OCJOaVI8ZmZWsn57/BHxKrBLGuIxM7M2UM8Y/xLgbkk3U/EzjBHxrbKCMjOz8vTZ45f0g3T3BOBnqez2FTczMxuB+uvxHyjpTRRTMF/SpHjMzKxk/SX+f6SYknkPYG7FeuF5+M3MRqw+h3oi4tsR8Vbg+xGxZ8XN8/CbmY1gA17HHxGfaEYgZmbWHHVN2WBmZu3Did/MLDNO/GZmmXHiNzPLTGmJX9IVklZLWlCxbidJt0t6OP3dsaz6zcystjJ7/FcCR1etOwe4IyL2Au5Iy2Zm1kSlJf6ImAOsqVp9DHBVun8V8MGy6jczs9oUEeUdXOoEfhYR+6XlZyJifMX2tRFRc7hH0jRgGkBHR8eBs2bNaiiG1WvWseqFhnZtSx2jcXskbove3B69DZf2mDxpXMP7Tp06dV5EdFWvr2d2zpaIiMuAywC6urqiu7u7oeNccs1NzJw/bP/Npps+eaPbI3Fb9Ob26G24tMeSU7qH/JjNvqpnlaQJAOnv6ibXb2aWvWYn/puB09P904Gbmly/mVn2yryc81rgt8A+kpZLOguYARwl6WHgqLRsZmZNVNoAVkSc1MemI8qq08zMBuZv7pqZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmXHiNzPLjBO/mVlmnPjNzDLjxG9mlhknfjOzzDjxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8vMlq2oVNIS4DngVWBjRHS1Ig4zsxy1JPEnUyPiqRbWb2aWJQ/1mJllRhHR/Eqlx4C1QAD/FBGX1SgzDZgG0NHRceCsWbMaqmv1mnWsemEQwbaZjtG4PRK3RW9uj96GS3tMnjSu4X2nTp06r9ZQeqsS/8SIeELSrsDtwKcjYk5f5bu6umLu3LkN1XXJNTcxc34rR7SGl+mTN7o9ErdFb26P3oZLeyyZ8b6G95VUM/G3ZKgnIp5If1cDNwAHtSIOM7McNT3xSxojafue+8C7gQXNjsPMLFeteB/TAdwgqaf+H0bEL1sQh5lZlpqe+CNiMXBAs+s1M7OCL+c0M8uME7+ZWWac+M3MMuPEb2aWGSd+M7PMOPGbmWXGid/MLDNO/GZmmXHiNzPLjBO/mVlmnPjNzDLjxG9mlhknfjOzzDjxm5llxonfzCwzTvxmZplx4jczy4wTv5lZZpz4zcwy48RvZpYZJ34zs8w48ZuZZcaJ38wsM078ZmaZceI3M8uME7+ZWWac+M3MMuPEb2aWmZYkfklHS3pI0iOSzmlFDGZmuWp64pc0CvgO8F5gX+AkSfs2Ow4zs1y1osd/EPBIRCyOiJeBWcAxLYjDzCxLiojmVigdBxwdER9Ny6cCfxoRn6oqNw2Ylhb3AR5qsMqdgaca3LcduT02cVv05vborR3a400RsUv1yi1bEIhqrHvdq09EXAZcNujKpLkR0TXY47QLt8cmbove3B69tXN7tGKoZzmwW8XyG4EnWhCHmVmWWpH4fw/sJWkPSVsDJwI3tyAOM7MsNX2oJyI2SvoUcBswCrgiIhaWWOWgh4vajNtjE7dFb26P3tq2PZr+4a6ZmbWWv7lrZpYZJ34zs8yM+MQv6QpJqyUtqFr/6TQtxEJJ36hYf26aKuIhSe9pfsTl2Zy2kHSUpHmS5qe/h7cm6vJs7rmRtu0uab2kzzc32vI18FzZX9Jv0/r5krZtftTl2MznylaSrkptsEjSua2JeghFxIi+AYcB/x1YULFuKvArYJu0vGv6uy9wP7ANsAfwKDCq1f9Di9riT4CJ6f5+wIpWx9/K9qjYfh3wE+DzrY6/xefHlsADwAFp+Q0ZP1dOBmal+9sBS4DOVv8Pg7mN+B5/RMwB1lSt/gQwIyJeSmVWp/XHUDyAL0XEY8AjFFNItIXNaYuI+I+I6Pn+xEJgW0nbNC3YJtjMcwNJHwQWU7RH29nM9ng38EBE3J/WPx0RrzYt2JJtZlsEMEbSlsBo4GXg2WbFWoYRn/j7sDdwqKTfSbpL0tvT+knAsopyy9O6dtZXW1T6C+A/ek74NlezPSSNAb4AfKml0TVfX+fH3kBIuk3SvZLObmGMzdJXW/wU2ACsBJYCF0ZE9YvGiNKKKRuaYUtgR+DPgLcDP5a0J3VOF9FmarZFpPetkt4G/D1FDy8HfZ0bXwIuioj1Uq3TpG311R5bAoekdc8Dd0iaFxF3tCzS8vXVFgcBrwIT0/ZfS/pVRCxuWaSD1K6JfzlwfUpu90h6jWLCpRyni+irLZ6U9EbgBuC0iHi0lUE2UV/t8afAcekDvfHAa5JejIhLWxdqU/T3XLkrIp4CkHQrxZh4Oyf+vtriZOCXEfEKsFrS3UAXxbDgiNSuQz03AocDSNob2Jpilr2bgRMlbSNpD2Av4J5WBdkkN1KjLSSNB34OnBsRd7csuua7kRrtERGHRkRnRHQCFwNfyyDpQ9/PlduA/SVtl8a23wU82Kogm+RGarfFUuBwFcZQvCP4Q6uCHBKt/nR5sDfgWoqxt1coXrHPonjArgYWAPcCh1eU/1uKq3keAt7b6vhb1RbAFynGLe+ruO3ayvhbfW5U7Hc+7XlVz+Y+Vz5M8UH3AuAbrY6/VW0BjKW40mshxYvf37Q6/sHePGWDmVlm2nWox8zM+uDEb2aWGSd+M7PMOPGbmWXGid/MLDNO/NZ2JI2X9Fd1lOuUdHKd5RbUUe5KScfVG2fVvt2SDh6KY5kNxInf2tF4YMDED3RSfCtzOOgGDh6okNlQcOK3djQDeLOk+yR9M33j8puSFqQ51U+oKHdoKve51LP/dZqU7N7KHngt6biXSnpQ0s+BXSu2HZgm+pqXJjqbkNbPlnSxpN+keA6S1Al8HPhciuXQdJjDUrnF7v3bUGrXuXosb+cA+0XEFABJfwFMAQ6gmHvl95LmpHKfj4j3p3LbAUdFxIuS9qL4dmdXP/UcC+wDTAY6KL7VeYWkrYBLgGMi4sn0QvNV4My035iIOFjSYcAVEbGfpH8E1kfEhSmWs4AJFBOlvYViupGfDr5pzJz4LQ+HANdGMZ/8Kkl3Ucy+WD2n+lbApZKmUMzGuPcAxz2s4rhPSPrXtH4fih+3uT3N9DmKYnqAHtdCMSe8pB3SvEm13BgRrwEPSuoY+N80q48Tv+Wg3nmWPwesonhnsAXwYh371JrzRMDCiHhHnfv0NW9K5e8jZDVXtJXLY/zWjp4Dtq9YngOcIGmUpF0oeur31Cg3DliZetmnUvTU+zOHYrbXUWkMf2pa/xCwi6R3wH/9ZuvbKvY7Ia0/BFgXEetqxGJWGvf4re1ExNOS7k6XYP4COBt4B8XvLQdwdkT8UdLTwEZJ9wNXAt8FrpN0PHAnxeyl/bmBYhrf+cB/Anel+l9OH8Z+W9I4iufZxWz6Sce1kn4D7MCmcf9bgJ9KOgb49CCbwKxfnp3TrIkkzab4QHluq2OxfHmox8wsM+7xm5llxj1+M7PMOPGbmWXGid/MLDNO/GZmmXHiNzPLzP8HoNjYxoNVjVwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select only the total depth vaues < 170 using conitional subsetting on the pandas series\n",
    "# the resulting data structure is still a pandas Series so we can plot the histogram\n",
    "# usingthe hist method\n",
    "# we see a close to uniform distribution for depth values < 170\n",
    "# I also added here a couple of ways we can customize the histogram\n",
    "\n",
    "hist_plt = dp_data[dp_data<170].hist()\n",
    "# dir(hist_plt)\n",
    "hist_plt.set_xlabel(\"total depth\")\n",
    "hist_plt.set_ylabel(\"frequency\")\n",
    "title_obj = hist_plt.set_title(\"total depth < 170 histogram\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "..................................................................................................\n",
    "\n",
    "<b><font color = \"red\">Exercise</font> <br></b>\n",
    "\n",
    "\n",
    "<b>Quality and SNP distribution</b> \n",
    "\n",
    "<b>A.</b> Display the distribution of the quality scores for the variants. \n",
    "- Select the QUAL column from the data and plot the histogram\n",
    "\n",
    "<b>B.</b> Compute the frequency (number of occurences) of each of the REF ALT combinations available in the dataset and make a bar plot.    \n",
    "Then, identify the most frequent combination of REF ALT in the dataframe \n",
    "- Compute the number of occurences for REF ALT combinations. \n",
    "    - Make use of the .value_counts() method for a dataframe to count the number for the unique combinations of row elements.\n",
    "- The result of value_counts is a pandas series, plot the values using a barplot.\n",
    "- Select from the series the maximum value together with the associated label: e.g. A T 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAATiklEQVR4nO3db4ydZ5nf8e+PJBvSeEuSBkbGceus1rvdJBahjFK29MWY0E0WVjVITWuUZZ3CyrwIKrSuWpu+AIQs5QWGrgRBNYRiFZZZK0BjJZvdZl1GLNKGELPZdZxgYRE36yS1d5cQmAhFHXP1xTwRZ50Zz3jOmTMz9/l+pKNznvv5c67LZ/ybZ+55zplUFZKktrxqpQuQJA2e4S5JDTLcJalBhrskNchwl6QGXbzSBQBcffXVtWnTppUuYyhefPFFLr/88pUuY6hGsWcYzb5HsWdYub6PHDnyN1X12rnWrYpw37RpE48++uhKlzEUU1NTTExMrHQZQzWKPcNo9j2KPcPK9Z3k/8y3zmkZSWqQ4S5JDTLcJalBhrskNchwl6QGLRjuSV6d5JEkf5HkWJKPdeMfTfJMkse629t79tmT5ESS40luWc4GJEmvtJhLIV8C3lpV00kuAb6V5MFu3aeq6hO9Gye5DtgOXA+8HviTJL9SVWcHWbgkaX4LnrnXrOlu8ZLudr7PCd4GTFbVS1X1FHACuKnvSiVJi7aoNzEluQg4Avwy8Jmq+naS3wQ+kOR3gEeBXVX1PLABeLhn91Pd2LnH3AnsBBgbG2NqaqqfPtaM6enpken1ZaPYM4xm36PYM6zOvhcV7t2Uyo1JrgC+nuQG4LPAx5k9i/84sA94L5C5DjHHMfcD+wHGx8drVN7Vttbfwbdp9wMXvM+uLWfZ960X+3rek3e9o6/9V8Jaf62XYhR7htXZ9wVdLVNVPwKmgFur6nRVna2qnwGf4+dTL6eAjT27XQM823+pkqTFWszVMq/tzthJchnwNuB7Sdb3bPYu4PHu8SFge5JLk1wLbAYeGWjVkqTzWsy0zHrgQDfv/irgYFXdn+R/JLmR2SmXk8D7AarqWJKDwBPADHCnV8pI0nAtGO5V9ZfAG+cYf8959tkL7O2vNEnSUvkOVUlqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktSgVfEHsnVhlvIRAJJGi2fuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBC4Z7klcneSTJXyQ5luRj3fhVSR5K8v3u/sqeffYkOZHkeJJblrMBSdIrLebM/SXgrVX1BuBG4NYkbwZ2A4erajNwuFsmyXXAduB64Fbg7iQXLUPtkqR5LBjuNWu6W7ykuxWwDTjQjR8A3tk93gZMVtVLVfUUcAK4aZBFS5LOL1W18EazZ95HgF8GPlNV/znJj6rqip5tnq+qK5N8Gni4qr7Ujd8DPFhV955zzJ3AToCxsbE3TU5ODqqnVW16epp169b1dYyjz7wwoGqGY+wyOP3T/o6xZcNrBlPMEA3itV5rRrFnWLm+t27deqSqxudat6g/s1dVZ4Ebk1wBfD3JDefZPHMdYo5j7gf2A4yPj9fExMRiSlnzpqam6LfXO9bYn9nbtWWGfUf7+4uOJ2+fGEwxQzSI13qtGcWeYXX2fUFXy1TVj4ApZufSTydZD9Ddn+k2OwVs7NntGuDZfguVJC3eYq6WeW13xk6Sy4C3Ad8DDgE7us12APd1jw8B25NcmuRaYDPwyIDrliSdx2J+Vl4PHOjm3V8FHKyq+5P8GXAwyfuAp4HbAKrqWJKDwBPADHBnN60jSRqSBcO9qv4SeOMc438L3DzPPnuBvX1XJ0laEt+hKkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBi0Y7kk2JvlGkieTHEvywW78o0meSfJYd3t7zz57kpxIcjzJLcvZgCTplS5exDYzwK6q+m6SXwSOJHmoW/epqvpE78ZJrgO2A9cDrwf+JMmvVNXZQRYuSZrfgmfuVfVcVX23e/wT4Elgw3l22QZMVtVLVfUUcAK4aRDFSpIWJ1W1+I2TTcA3gRuA/wDcAfwYeJTZs/vnk3waeLiqvtTtcw/wYFXde86xdgI7AcbGxt40OTnZdzNrwfT0NOvWrevrGEefeWFA1QzH2GVw+qf9HWPLhtcMppghGsRrvdaMYs+wcn1v3br1SFWNz7VuMdMyACRZB3wV+FBV/TjJZ4GPA9Xd7wPeC2SO3V/xHaSq9gP7AcbHx2tiYmKxpaxpU1NT9NvrHbsfGEwxQ7Jrywz7ji76S21OJ2+fGEwxQzSI13qtGcWeYXX2vairZZJcwmywf7mqvgZQVaer6mxV/Qz4HD+fejkFbOzZ/Rrg2cGVLElayGKulglwD/BkVX2yZ3x9z2bvAh7vHh8Ctie5NMm1wGbgkcGVLElayGJ+Vn4L8B7gaJLHurEPA+9OciOzUy4ngfcDVNWxJAeBJ5i90uZOr5SRpOFaMNyr6lvMPY/+h+fZZy+wt4+6JEl98B2qktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ1aMNyTbEzyjSRPJjmW5IPd+FVJHkry/e7+yp599iQ5keR4kluWswFJ0ist5sx9BthVVb8GvBm4M8l1wG7gcFVtBg53y3TrtgPXA7cCdye5aDmKlyTNbcFwr6rnquq73eOfAE8CG4BtwIFuswPAO7vH24DJqnqpqp4CTgA3DbhuSdJ5pKoWv3GyCfgmcAPwdFVd0bPu+aq6MsmngYer6kvd+D3Ag1V17znH2gnsBBgbG3vT5ORkn62sDdPT06xbt66vYxx95oUBVTMcY5fB6Z/2d4wtG14zmGKGaBCv9Vozij3DyvW9devWI1U1Pte6ixd7kCTrgK8CH6qqHyeZd9M5xl7xHaSq9gP7AcbHx2tiYmKxpaxpU1NT9NvrHbsfGEwxQ7Jrywz7ji76S21OJ2+fGEwxQzSI13qtGcWeYXX2vairZZJcwmywf7mqvtYNn06yvlu/HjjTjZ8CNvbsfg3w7GDKlSQtxmKulglwD/BkVX2yZ9UhYEf3eAdwX8/49iSXJrkW2Aw8MriSJUkLWczPym8B3gMcTfJYN/Zh4C7gYJL3AU8DtwFU1bEkB4EnmL3S5s6qOjvowiVJ81sw3KvqW8w9jw5w8zz77AX29lGXJKkPvkNVkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUELhnuSLyQ5k+TxnrGPJnkmyWPd7e096/YkOZHkeJJblqtwSdL8FnPm/kXg1jnGP1VVN3a3PwRIch2wHbi+2+fuJBcNqlhJ0uIsGO5V9U3gh4s83jZgsqpeqqqngBPATX3UJ0lagov72PcDSX4HeBTYVVXPAxuAh3u2OdWNvUKSncBOgLGxMaampvooZe2Ynp7uu9ddW2YGU8yQjF3Wf81r8etjEK/1WjOKPcPq7Hup4f5Z4ONAdff7gPcCmWPbmusAVbUf2A8wPj5eExMTSyxlbZmamqLfXu/Y/cBgihmSXVtm2He0n/MIOHn7xGCKGaJBvNZrzSj2DKuz7yVdLVNVp6vqbFX9DPgcP596OQVs7Nn0GuDZ/kqUJF2oJYV7kvU9i+8CXr6S5hCwPcmlSa4FNgOP9FeiJOlCLfizcpKvABPA1UlOAR8BJpLcyOyUy0ng/QBVdSzJQeAJYAa4s6rOLkvlkqR5LRjuVfXuOYbvOc/2e4G9/RQlSeqP71CVpAb1dwmDNCSbVvAKoZN3vWPFnltaKs/cJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1aMFwT/KFJGeSPN4zdlWSh5J8v7u/smfdniQnkhxPcstyFS5Jmt9izty/CNx6zthu4HBVbQYOd8skuQ7YDlzf7XN3kosGVq0kaVEWDPeq+ibww3OGtwEHuscHgHf2jE9W1UtV9RRwArhpMKVKkhbr4iXuN1ZVzwFU1XNJXteNbwAe7tnuVDf2Ckl2AjsBxsbGmJqaWmIpa8v09HTfve7aMjOYYoZk7LK1V3Ovpb5eg3it15pR7BlWZ99LDff5ZI6xmmvDqtoP7AcYHx+viYmJAZeyOk1NTdFvr3fsfmAwxQzJri0z7Ds66C+14Tl5+8SS9hvEa73WjGLPsDr7XurVMqeTrAfo7s9046eAjT3bXQM8u/TyJElLsdRwPwTs6B7vAO7rGd+e5NIk1wKbgUf6K1GSdKEW/Fk5yVeACeDqJKeAjwB3AQeTvA94GrgNoKqOJTkIPAHMAHdW1dllql2SNI8Fw72q3j3Pqpvn2X4vsLefoiRJ/fEdqpLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGrTgn9nT/DbtfuCC99m1ZYY7lrCfJF0Iz9wlqUGGuyQ1qK9pmSQngZ8AZ4GZqhpPchXwB8Am4CTwr6vq+f7KlCRdiEGcuW+tqhurarxb3g0crqrNwOFuWZI0RMsxLbMNONA9PgC8cxmeQ5J0Hqmqpe+cPAU8DxTw36pqf5IfVdUVPds8X1VXzrHvTmAnwNjY2JsmJyeXXMdKOfrMCxe8z9hlcPqny1DMKjaKPUP/fW/Z8JrBFTMk09PTrFu3bqXLGLqV6nvr1q1HemZN/o5+L4V8S1U9m+R1wENJvrfYHatqP7AfYHx8vCYmJvosZfiWcknjri0z7Ds6WlegjmLP0H/fJ2+fGFwxQzI1NcVa/L/cr9XYd1/TMlX1bHd/Bvg6cBNwOsl6gO7+TL9FSpIuzJLDPcnlSX7x5cfAbwCPA4eAHd1mO4D7+i1SknRh+vlZeQz4epKXj/P7VfVHSb4DHEzyPuBp4Lb+y5QkXYglh3tV/QB4wxzjfwvc3E9RkqT++A5VSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDRu+vFktrxKYl/AH2QTl51ztW7Lk1GJ65S1KDmjhzX8kzHElajTxzl6QGGe6S1KBlm5ZJcivwe8BFwOer6q7lei5Jg7XUqc5dW2a4o49pUn+ROzjLEu5JLgI+A/wL4BTwnSSHquqJ5Xg+SW1Yqd+ftfhNZbnO3G8CTlTVDwCSTALbAMNd0qrT7zeVfn5iWa5vLKmqwR80+VfArVX1u93ye4B/WlUf6NlmJ7CzW/xV4PjAC1mdrgb+ZqWLGLJR7BlGs+9R7BlWru9/VFWvnWvFcp25Z46xv/NdpKr2A/uX6flXrSSPVtX4StcxTKPYM4xm36PYM6zOvpfraplTwMae5WuAZ5fpuSRJ51iucP8OsDnJtUl+AdgOHFqm55IknWNZpmWqaibJB4A/ZvZSyC9U1bHleK41aOSmohjNnmE0+x7FnmEV9r0sv1CVJK0s36EqSQ0y3CWpQYb7MkmyMck3kjyZ5FiSD3bjVyV5KMn3u/srV7rWQUtyUZI/T3J/tzwKPV+R5N4k3+te819vve8k/7772n48yVeSvLrFnpN8IcmZJI/3jM3bZ5I9SU4kOZ7klpWp2nBfTjPArqr6NeDNwJ1JrgN2A4erajNwuFtuzQeBJ3uWR6Hn3wP+qKr+MfAGZvtvtu8kG4B/B4xX1Q3MXjixnTZ7/iJw6zljc/bZ/R/fDlzf7XN393Esw1dV3oZwA+5j9rN2jgPru7H1wPGVrm3AfV7D7Bf7W4H7u7HWe/77wFN0Fyj0jDfbN7AB+CvgKmavursf+I1WewY2AY8v9NoCe4A9Pdv9MfDrK1GzZ+5DkGQT8Ebg28BYVT0H0N2/bgVLWw7/FfhPwM96xlrv+ZeAvwb+ezcd9fkkl9Nw31X1DPAJ4GngOeCFqvpfNNzzOebr8+Vvei871Y0NneG+zJKsA74KfKiqfrzS9SynJL8FnKmqIytdy5BdDPwT4LNV9UbgRdqYjphXN8e8DbgWeD1weZLfXtmqVoUFP3plWAz3ZZTkEmaD/ctV9bVu+HSS9d369cCZlapvGbwF+JdJTgKTwFuTfIm2e4bZs7NTVfXtbvleZsO+5b7fBjxVVX9dVf8P+Brwz2i7517z9blqPnrFcF8mSQLcAzxZVZ/sWXUI2NE93sHsXHwTqmpPVV1TVZuY/aXS/66q36bhngGq6v8Cf5XkV7uhm5n9eOuW+34aeHOSv9d9rd/M7C+RW+6513x9HgK2J7k0ybXAZuCRFajPd6gulyT/HPhT4Cg/n3/+MLPz7geBf8jsf5DbquqHK1LkMkoyAfzHqvqtJP+AxntOciPweeAXgB8A/5bZk6dm+07yMeDfMHtl2J8Dvwuso7Gek3wFmGD2Y31PAx8B/ifz9JnkvwDvZfbf5UNV9eDwqzbcJalJTstIUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktSg/w9ar/qOVCAI4AAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# A. Display the distribution of the quality scores for the variants.\n",
    "# Select the QUAL column from the data and plot the histogram\n",
    "\n",
    "vcf_data.QUAL.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "REF  ALT\n",
       "C    T      175\n",
       "G    A      163\n",
       "A    G      155\n",
       "T    C      124\n",
       "C    A       56\n",
       "G    T       53\n",
       "C    G       49\n",
       "G    C       46\n",
       "T    G       43\n",
       "     A       35\n",
       "A    C       34\n",
       "     T       26\n",
       "dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# B. Compute the frequency (number of occurences) of each of the REF ALT \n",
    "# combinations available in the dataset and make a bar plot.\n",
    "# Then, identify the most frequent combination of REF ALT in the dataframe\n",
    "\n",
    "# Compute the number of occurences for REF ALT combinations.\n",
    "# Make use of the .value_counts() method for a dataframe \n",
    "# to count the number for the unique combinations of row elements.\n",
    "\n",
    "\n",
    "REF_ALT_counts = vcf_data.loc[:,(\"REF\",\"ALT\")].value_counts()\n",
    "REF_ALT_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we have a series, the row labels look like two columns but they are just row labels\n",
    "type(REF_ALT_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex([('C', 'T'),\n",
       "            ('G', 'A'),\n",
       "            ('A', 'G'),\n",
       "            ('T', 'C'),\n",
       "            ('C', 'A'),\n",
       "            ('G', 'T'),\n",
       "            ('C', 'G'),\n",
       "            ('G', 'C'),\n",
       "            ('T', 'G'),\n",
       "            ('T', 'A'),\n",
       "            ('A', 'C'),\n",
       "            ('A', 'T')],\n",
       "           names=['REF', 'ALT'])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "REF_ALT_counts.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='REF,ALT'>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEZCAYAAABsPmXUAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAWZUlEQVR4nO3df7BkZX3n8fcHEMwiUWCuFOGHFxDcUkzGdWQ3UVOAa8QfUclqwsQoRpIRC9fdZJMNulYkW7prxV/ZXRFqjERMFFAJ0QT8gUYlbhJ10BFHhPDDUQcIM8JGTQVNwO/+0WfK9nov90736XPnPrxfVbem+zmnz/fpPs987rmnTz+dqkKS1JZ9VrsDkqT+Ge6S1CDDXZIaZLhLUoMMd0lqkOEuSQ3ab7U7ALBu3bqan59f7W5I0ppy7bXXfrOq5hZbtleE+/z8PFu2bFntbkjSmpLka0st87SMJDXIcJekBhnuktQgw12SGmS4S1KDlg33JBcl2Zlk21jbZUm2dj/bk2zt2ueT3DO27MIZ9l2StISVXAr5TuCtwLt2N1TVL+2+neRNwLfG1r+lqtb31D9J0gSWDfequibJ/GLLkgT4ReDUnvslSZrCtB9iejJwZ1XdNNZ2TJIvAN8GXl1Vf7XYA5NsAjYBHH300UsWmD/3yok6tv31z5zocZLUgmnfUN0IXDJ2/w7g6Kp6HPCbwHuS/PhiD6yqzVW1oao2zM0t+ulZSdKEJg73JPsBvwBctrutqr5XVXd1t68FbgFOmLaTkqQ9M82R+78HbqiqHbsbkswl2be7fSxwPHDrdF2UJO2plVwKeQnwN8CjkuxIcla36Ax++JQMwM8C1yX5IvB+4OyqurvPDkuSlreSq2U2LtH+4kXaLgcun75bkqRp+AlVSWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoOmnRWyOc5CKakFHrlLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KD/ITqKvMTsZJmwSN3SWrQsuGe5KIkO5NsG2s7L8ltSbZ2P88YW/bKJDcnuTHJ02bVcUnS0lZy5P5O4LRF2t9SVeu7n6sAkjwaOAN4TPeYtyXZt6/OSpJWZtlwr6prgLtXuL3nAJdW1feq6qvAzcBJU/RPkjSBac65vzzJdd1pm4O7tiOAb4yts6NrkyQNaNJwvwA4DlgP3AG8qWvPIuvWYhtIsinJliRbdu3aNWE3JEmLmSjcq+rOqrqvqr4PvJ0fnHrZARw1tuqRwO1LbGNzVW2oqg1zc3OTdEOStISJwj3J4WN3Twd2X0nzQeCMJAckOQY4HvjsdF2UJO2pZT/ElOQS4GRgXZIdwGuAk5OsZ3TKZTvwUoCq+nKS9wLXA/cC51TVfTPpuSRpScuGe1VtXKT5Hfez/uuA103TKUnSdPyEqiQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDlg33JBcl2Zlk21jbG5LckOS6JFckeVjXPp/kniRbu58LZ9h3SdISVnLk/k7gtAVtVwMnVtVPAn8HvHJs2S1Vtb77ObufbkqS9sSy4V5V1wB3L2j7aFXd2939W+DIGfRNkjShPs65vwT40Nj9Y5J8Icmnkjx5qQcl2ZRkS5Itu3bt6qEbkqTd9pvmwUn+G3Av8O6u6Q7g6Kq6K8njgT9L8piq+vbCx1bVZmAzwIYNG2qafmjl5s+9cqLHbX/9M3vuiaRZmvjIPcmZwLOAF1RVAVTV96rqru72tcAtwAl9dFSStHIThXuS04DfAZ5dVf801j6XZN/u9rHA8cCtfXRUkrRyy56WSXIJcDKwLskO4DWMro45ALg6CcDfdlfG/Czw35PcC9wHnF1Vdy+6YUnSzCwb7lW1cZHmdyyx7uXA5dN2SpI0HT+hKkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBi0b7kkuSrIzybaxtkOSXJ3kpu7fg8eWvTLJzUluTPK0WXVckrS0lRy5vxM4bUHbucDHq+p44OPdfZI8GjgDeEz3mLcl2be33kqSVmTZcK+qa4C7FzQ/B7i4u30x8Nyx9kur6ntV9VXgZuCkfroqSVqpSc+5H1ZVdwB0/z68az8C+MbYeju6NknSgPp+QzWLtNWiKyabkmxJsmXXrl09d0OSHtgmDfc7kxwO0P27s2vfARw1tt6RwO2LbaCqNlfVhqraMDc3N2E3JEmLmTTcPwic2d0+E/jAWPsZSQ5IcgxwPPDZ6booSdpT+y23QpJLgJOBdUl2AK8BXg+8N8lZwNeB5wNU1ZeTvBe4HrgXOKeq7ptR3yVJS1g23Ktq4xKLnrLE+q8DXjdNpyRJ0/ETqpLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkN2m/SByZ5FHDZWNOxwO8CDwN+HdjVtb+qqq6atI4kac9NHO5VdSOwHiDJvsBtwBXArwJvqao39tFBSdKe6+u0zFOAW6rqaz1tT5I0hb7C/QzgkrH7L09yXZKLkhy82AOSbEqyJcmWXbt2LbaKJGlCU4d7kv2BZwPv65ouAI5jdMrmDuBNiz2uqjZX1Yaq2jA3NzdtNyRJY/o4cn868PmquhOgqu6sqvuq6vvA24GTeqghSdoDfYT7RsZOySQ5fGzZ6cC2HmpIkvbAxFfLACT5V8BTgZeONf9+kvVAAdsXLJMkDWCqcK+qfwIOXdD2wql6JEmamp9QlaQGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAZN9U1M0nLmz71yosdtf/0ze+6J9MDikbskNchwl6QGTXVaJsl24DvAfcC9VbUhySHAZcA8sB34xar6f9N1U1oZTwNJI30cuZ9SVeurakN3/1zg41V1PPDx7r4kaUCzOC3zHODi7vbFwHNnUEOSdD+mDfcCPprk2iSburbDquoOgO7fh09ZQ5K0h6a9FPKJVXV7kocDVye5YaUP7H4ZbAI4+uijp+yGJGncVOFeVbd3/+5McgVwEnBnksOr6o4khwM7l3jsZmAzwIYNG2qafkirxTdwtbea+LRMkgOTHLT7NvBzwDbgg8CZ3WpnAh+YtpOSpD0zzZH7YcAVSXZv5z1V9eEknwPem+Qs4OvA86fvpiRpT0wc7lV1K/BTi7TfBTxlmk5Jkqbj3DLSGuI5fq2U0w9IUoMMd0lqkKdlJC1pktNAngLaO3jkLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgwx3SWqQ4S5JDfITqpL2Ck6K1i/DXdIDUuu/TDwtI0kNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkho0cbgnOSrJJ5J8JcmXk/ynrv28JLcl2dr9PKO/7kqSVmKa69zvBf5LVX0+yUHAtUmu7pa9pareOH33JEmTmDjcq+oO4I7u9neSfAU4oq+OSVJLhv7QVC/n3JPMA48DPtM1vTzJdUkuSnJwHzUkSSs3dbgneQhwOfCfq+rbwAXAccB6Rkf2b1ricZuSbEmyZdeuXdN2Q5I0ZqpwT/IgRsH+7qr6U4CqurOq7quq7wNvB05a7LFVtbmqNlTVhrm5uWm6IUlaYJqrZQK8A/hKVb15rP3wsdVOB7ZN3j1J0iSmuVrmicALgS8l2dq1vQrYmGQ9UMB24KVT1JAkTWCaq2U+DWSRRVdN3h1JUh/8hKokNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSg2YW7klOS3JjkpuTnDurOpKkHzWTcE+yL3A+8HTg0cDGJI+eRS1J0o+a1ZH7ScDNVXVrVf0zcCnwnBnVkiQtkKrqf6PJ84DTqurXuvsvBP5tVb18bJ1NwKbu7qOAGycotQ745pTdtZ71rLd317Le0h5RVXOLLdhvuv4sKYu0/dBvkaraDGyeqkiypao2TLMN61nPent3LetNZlanZXYAR43dPxK4fUa1JEkLzCrcPwccn+SYJPsDZwAfnFEtSdICMzktU1X3Jnk58BFgX+CiqvryDEpNdVrHetaz3pqoZb0JzOQNVUnS6vITqpLUIMNdkhpkuEtSg2Z1nfvMJTkQ+G5V3TfDGkcyutLnycBPAPcA24ArgQ9V1ffXcr2xukO8lj8N/Aqj53Y4P/zc/qSqvrXG6zU9VoZ+PRfUbm58Lqg9k+e3Zt5QTbIPo8H8AuAJwPeAA4BdwFXA5qq6qcd6fwQcAfwFsAXYCTwYOAE4BXg8cG5VXbPW6q3Ca/khRp9z+ACLP7efB95cVb1cLrsK9ZodK129oV/P1sfnIM9vLYX7p4CPMdoB23YfmSQ5hNEO+GXgiqr6k57qnVhV2+5n+f7A0VV181qrtwqv5bqqut+PVq9knb24XrNjpdve0K9n6+NzkOe3lsL9QVX1L9Ou00M/jgLOqKo3zLLOLO1Fr+UTgV+uqnN63u5Hq+rn+tzm3iTJHDBXVdcvaH8MsLOqdq1Oz/ox9PhM8kjgsKr6vwvanwzcXlW39FFnbLuDPL+19IbqZ5ZbYVZhlGRdkpcluQb4JHDYDGo8J8k5Y/c/k+TW7uf5PZdbn+Tpi/Th2UkeDzN9Ldcn+f0k24HXAjfMoMyiEynNysD7DuD/sPhzPBL4X30XS3JWkt8eu39bkm8n+U6Sl/Vdb7Gxl+TAJL+S5Mql1pnCHwDfWaT9nm5Z3wbJsrX0hupik5HNrlhyEHA6oz+RTgCuAI6tqiNnVPK/MjoPt9sBjM7HHQj8EfC+Hmu9AXjxIu3XM/qk3Kk91iLJCYye20bgLuAyRn81ntJnnTEPTfILSy2sqj/tud6Q+w7gsVX1qYWNVfWRJG/quRbA2cBpY/d3VtURSR4MfBS4YAY1d59eegaj/4OnAZcDF86g1HxVXbewsaq2JJmfQb1Bsmwthftckt9camFVvbnnejuBzwKvBj5dVZXk9J5rjNu/qr4xdv/TVXUXcFf3bnqfDq2q7Qsbq+rmJIf2XAtGR+d/Bfz87vPASX5jBnV2eyjwLJaenbTvcB9y3wE8aMJlk9qnez67vQ+gqr6b5Mf6LpbkqYwOBJ4GfAL4Y+CkqvrVvmt1Hnw/y3p/fgyUZWsp3PcFHsJwR/CvYnQ0dgHwniSXzbjeweN3xue+p//TDPc3YGcRRv+B0Wv5iSQfZvTlLbPcj1+rqpfMcPsLDbnvAG5K8oyqumq8sTvVdusM6j10/E5V/Y+u3j7ALA4GPsLoYOBJVfXVrlbvp5vGfC7Jr1fV28cbk5wFXDuDeoNk2Vp6Q/XzVfVvVqHusYyOIs4Ajgdew+id7L/ruc67gU8uMsBeCpxcVRt7rHUho9Mjr66xAZDk94DDq2rTkg+eru6BwHMZvZ6nAhczei0/2nOdL1TV4/rc5jL1Btt33XZPYHQZ5F/zg/DZAPw08KwZjM23AXdX1asXtL8WWFdVZ/dc73GM/r89j9Evq0uB362qR/RZZ6zeYYxOu/4zP/x67g+cXlV/33O9QbJsLYX7oP9hl+jDYxkF0y9V1XE9b/vhwJ8xuub1813z4xmdv31uVd3ZY60DgT9k9HWIW7vmn2J0je+vVdU/9lXrfvpwCPB8Rq9l3+f47/dSwb4Nue/Gah7A6Fz0iV3Tl4H3VNV3Z1Br93h5AvDFrnmQ8dJdUbWR0V9/WxkdDMxkxsYkpzD2elbVX86oziBZtpbC/ZCqunu1+zFrSU4FHtPdndkA62odu6DWLP6kf8AYct+thgXj5fq+LxFcpvY+wFMZXYY8q3Pvgxgqy9ZMuEuSVm4tXecuSVohw10PGEkuTnJBkhOXX1ta29Z8uCf5WJIPJXnWQPUMiJ4Mve+AtzKa0+OFA9UblGOzX6uQLb3WW/Pn3JP8BKMpOv9dVZ0/QL0nAEcz+lDF7wxQ72PAvwDnV9VfzLjWV7qb51fVW2dZq6s36L4b2pD7rqs39Ngcerw0PT77rrdmwj2NT5a0lFUYYId2ta7scZuD7rskT2I0VcS7uvvvBw7pFr92qKtYWv/lBbMZL3tLvQw8SWDf9dbSaZmhJ0t6UpIXjd1/f5K/7H56vS57GfsCp/YZDkmW/NBQVd01g/84g+474PcYXYO926OA3wbOYzQPTK+SzCV59CKLDga+3newDz02hx4vqzA+x2vPfJLAoeqtpXBfcrIk4CdnUG/QgBg3wAAbdNZEht93P77gr4SbquraGn15xUEzqNf0Ly+GHy9Dz+p5UJIXdVNjfBZ4JKO//I6rqt9aq/XW0twyQ0+WtGhAACT5n30Xy7CzUA49a+LQ++5h43eqavy5zuJIbOhZGgcdmww/XoauN/QkgYPUW0vhPvRkSQ8bvzNAQAw5wIaeNXHofXdDkmcu/PO9uwrhxhnUa/2X19DjZeh6Q08SOEi9tfSG6tCTJf05cOESAfGyqnpmz/V+g9EOPxB4D6M5z6+uqmP7rNPVGnQStlXYd49k9MXGf80Pz/XyMzOqdyWjKzgW++X1iqr6kS9GmbLe0GNz6PHS9CSBQ9VbM+EOg0+WNGhAjNWd+QAbauKiBTUH23dj9V7A2Fwvs6r3APjlNeh4WY3xuUgfZjZJ4FD11ky4J0kt09mVrLOHNQcLiCXqz2SAJXlsVX1pmXV6ey2H3nerOFZa/eU19HhxfPZRr6rWxA+jq0b+I6NvdR9v358fzA3+4h7rpY919sZ6q/Batl6v2bHyANl/TdZbS0fuDwZewuho5RjgHxh9o9A+jL7H8fyq2tpjvU8y+s7GD1TV18fa9weeBJwJfKKq3rnW6i3xWj6Y0TX1s3gth953jpV+6+0N48V6e1pnrYT7uCQPAtYB91TVP8yoxt4QSDMbYGN1Z/5atl5vLxkrM6u3oHZz+6/Vemsy3IfW0g7XbDlWtLcw3CWpQWtp+gFJ0goZ7pLUIMNdTUpyX5KtSbYl+fMkD+va55Pc0y3b/fOibtn2JF8aa/+ZJbZ9epJK8q/H2uaTbOtuHzq2jb9PctvY/f0HePrSmppbRtoT91TVeoAkFwPnAK/rlt2ye9kiTqmqby6z7Y3Apxl9ivi8hQur6i5gd+3zgH+sqjfuUe+lKXnkrgeCvwGO6GNDSR4CPBE4i1G4S3slw11NS7Iv8BTgg2PNxy04LfPksWWf6No+s8Qmnwt8uEbzt9ydZPAJrqSV8LSMWvVjSbYC84wm87p6bNk0p2U2An/Q3b60u//5JdeWVonhrlbdU1XrkzyU0YyN5wD/e5oNZvT9nacCJyYpRp8eriQz/WYuaRKellHTqupbwCuA3+o+zbnHktzQ3Xwe8K6qekRVzVfVUcBXGc3nIu1VDHc1r6q+AHyRH7wBuvCc+yuWemySdfzgG4E2Mvr6w3GXM5rqF+BRSXaM/Ty/x6ch7RGnH5DuR0bfbnRsVU11SkcamuEuSQ3ytIwkNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ36/68tU32TqthCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot a bar plot with a bar for each value in the series\n",
    "\n",
    "REF_ALT_counts.plot(kind = \"bar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "REF  ALT\n",
       "C    T      175\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select from the series the maximum value together with the associated label: e.g. A T 300\n",
    "# get the label and the value from the series where the value is the maximum value\n",
    "# conitional subsetting\n",
    "\n",
    "REF_ALT_counts[REF_ALT_counts == max(REF_ALT_counts)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "REF  ALT\n",
       "C    T      175\n",
       "dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the data is sorted, descending, so take the first element, to add the label you need a range:\n",
    "\n",
    "REF_ALT_counts[0:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "175"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "REF_ALT_counts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "____\n",
    "____\n",
    "\n",
    "\n",
    "#### Read alignment analysis   \n",
    "We have the genome data/sequence for an organism and the sequence alingnment data for a (biological) sample from that organism.   \n",
    "We want to count how many reads align at each position in the genome.  \n",
    "This computation could be expanded further into the computation for creating a vcf file. \n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____\n",
    "\n",
    "<b><font color = \"red\">Exercise</font> <br></b>\n",
    "\n",
    "<b>Initialize the genome counts list</b>   \n",
    "\n",
    "Read the data from the the fasta file `test_genome.fa` and create/initialize a list for the genome.     \n",
    "Create a list where we have the value 0 for each position/nucleotide in the genome.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">I dna_rm:chromosome chromosome:R64-1-1:I:1:230218:1 REF\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# open the file and read the information line by line\n",
    "# where we have nucleotides ( ACGT or N = any) \n",
    "# we add a 0 in the genome positions list\n",
    "# because we do not have any reads aligning there yet\n",
    "\n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    for line in genome_file:\n",
    "        print(line)\n",
    "        break # stop to see what we have on the first line\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# the first line has some genime information and \n",
    "# no nucleotides so I can ignore it\n",
    "# either read it and move on\n",
    "# or check for lines that do not start with >\n",
    "\n",
    "\n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    genome_file.readline() # read and ignore the first line\n",
    "    for line in genome_file:\n",
    "        print(line)\n",
    "        break # stop to see what we have on the first line of nucleotides\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# or check for lines that do not start with >\n",
    "\n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    for line in genome_file:\n",
    "        if not line.startswith(\">\"):\n",
    "            print(line)\n",
    "            break # stop to see what we have on the first line of nucleotides\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create empty list of positions\n",
    "genome_pos = [] \n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    # read and ignore the first line because it does not have nucleotides\n",
    "    genome_file.readline()\n",
    "    for line in genome_file:\n",
    "        # go through each nucleotide one at a time, \n",
    "        # strip new line at the end of the line\n",
    "        for nc in line.strip():\n",
    "            # add a 0 for each nucleotide\n",
    "            genome_pos.append(0)\n",
    "        # stop to see what we have on the first line of nucleotides\n",
    "        # break "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list/genome size: 230218\n",
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "print(\"list/genome size:\" , len(genome_pos))\n",
    "print(genome_pos[:150])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "___\n",
    "\n",
    "<b> <font color = \"red\">Exercise</font></b>\n",
    "\n",
    "<b>Count the number of reads that align at each position in the genome.</b>    \n",
    "\n",
    "Read the data from the alignment file `sample.bam` using the bamnostic library.  \n",
    "##### To install the bamnostic package follow the instruction from the package github repository.\n",
    "https://github.com/betteridiot/bamnostic\n",
    "\n",
    "Go through each read and update the count in the list initialized in the previous exercise.   \n",
    "To update the count, add up 1 in the counts for each position the read covers.   \n",
    "The positions the read covers start with the read position and end with the read position + the length of the read sequence.       \n",
    "Consecutive nucleotides in the read algn to consecutive positions in the genome.   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your solution here\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To install the bamnostic package open a terminal and run:\n",
    "```bash\n",
    "pip install bamnostic\n",
    "```\n",
    "##### It is one of the install options provided in the documentation\n",
    "\n",
    "https://github.com/betteridiot/bamnostic\n",
    "\n",
    "If it does not work try another option\n",
    "\n",
    "There are 4 methods of installation available (choose one):\n",
    "\n",
    "Option 1: Through the conda package manager (Anaconda Cloud)\n",
    "\n",
    "```bash\n",
    "# first, add the conda-forge channel to your conda build\n",
    "conda config --add channels conda-forge\n",
    "\n",
    "# now bamnostic is available for install\n",
    "conda install bamnostic\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the library \n",
    "import bamnostic as bs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EAS56_57:6:190:289:82\t69\tchr1\t100\t0\t*\t=\t100\t0\tCTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA\t<<<7<<<;<<<<<<<<8;;<7;4<;<;;;;;94<;\tMF:C:192"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading your BAM file (Note: CRAM format are not supported at this time)\n",
    "# Bamnostic comes with an example BAM (and respective BAI) file just to play around with the output. \n",
    "# This example file is made availble through bamnostic.example_bam, which is a just a string path to the BAM file within the package.\n",
    "\n",
    "bam = bs.AlignmentFile(bs.example_bam, 'rb')\n",
    "\n",
    "# the result of the AlignmentFile function is a generator that gives us the aligned reads from the bam file one by one.\n",
    "first_read = next(bam)\n",
    "first_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dir(first_read)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the read sequence\n",
    "first_read.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the position where the read aligned\n",
    "# 0-based position, in the file we have 1-based positions \n",
    "# but for ease of preocessing they are set to 0-based in the read object\n",
    "\n",
    "first_read.pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99\n"
     ]
    }
   ],
   "source": [
    "print(first_read.pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA 99\n",
      "AGGGGTGCAGAGCCGAGTCACGGGGTTGCCAGCAC 99\n",
      "GGTGCAGAGCCGAGTCACGGGGTTGCCAGCACAGG 102\n",
      "AGCCGAGTCACGGGGTTGCCAGCACAGGGGCTTAA 109\n",
      "CCGAGTCACGGGGTTGCCAGCACAGGGGCTTAACC 111\n",
      "GTTCTCAAGGTTGTTGCAAGGGGGTCTATGTGAAC 112\n",
      "CGAGTCACGGGGTTGCCAGCACAGGGGCTTAACCT 112\n",
      "GTTCTCAAGGTTGTTGCAAGGGGGTCTATGTGAAC 118\n",
      "ACGGGGTTGCCAGCACAGGGGCTTAACCTCTGGTG 118\n",
      "GGGGTTGCCAGCACAGGGGCTTAACCTCTGGTGAC 120\n"
     ]
    }
   ],
   "source": [
    "# use a for loop to go through the first 10 reads\n",
    "# use a counter and break to stop after 10 reads\n",
    "# start from the beginning - recreate the generator\n",
    "\n",
    "bam = bs.AlignmentFile(bs.example_bam, 'rb')\n",
    "no_reads = 0\n",
    "for read in bam:\n",
    "    no_reads += 1 # printing another read so increase the number\n",
    "    print(read.seq, read.pos)\n",
    "    if no_reads >= 10: # when you reached 10 reads\n",
    "        break # exit the loop\n",
    "    \n",
    "# bam.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EAS192_3:1:225:195:543\t99\tchr1\t123\t99\t35M\t=\t299\t211\tGGTTGCCAGCACAGGGGCTTAACCTCTGGTGACTG\t<<<<<<<<<<<<<<<<<<<<<<<9<<;::388998\tAq:C:72\tH0:C:1\tH1:C:0\tMF:C:18\tNM:C:0\tUQ:C:0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting the next read because we did not close the file\n",
    "next(bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCAAGGTTGTTGCAAGGGGGTCTATGTGAACAAA 99\n",
      "AGGGGTGCAGAGCCGAGTCACGGGGTTGCCAGCAC 99\n",
      "GGTGCAGAGCCGAGTCACGGGGTTGCCAGCACAGG 102\n",
      "AGCCGAGTCACGGGGTTGCCAGCACAGGGGCTTAA 109\n",
      "CCGAGTCACGGGGTTGCCAGCACAGGGGCTTAACC 111\n",
      "GTTCTCAAGGTTGTTGCAAGGGGGTCTATGTGAAC 112\n",
      "CGAGTCACGGGGTTGCCAGCACAGGGGCTTAACCT 112\n",
      "GTTCTCAAGGTTGTTGCAAGGGGGTCTATGTGAAC 118\n",
      "ACGGGGTTGCCAGCACAGGGGCTTAACCTCTGGTG 118\n",
      "GGGGTTGCCAGCACAGGGGCTTAACCTCTGGTGAC 120\n"
     ]
    }
   ],
   "source": [
    "# Another option, using the with control structure so we close the file\n",
    "\n",
    "with bs.AlignmentFile(bs.example_bam, 'rb') as bam:\n",
    "    no_reads = 0\n",
    "    for read in bam:\n",
    "        no_reads += 1 # printing another read so increase the number\n",
    "        print(read.seq, read.pos)\n",
    "        if no_reads >= 10: # when you reached 10 reads\n",
    "            break # exit the loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "object of type 'NoneType' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/dg/8l7ql9hs6j71k502y3_6r3jh0000gp/T/ipykernel_7567/2408373372.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbam\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/bamnostic/bam.py\u001b[0m in \u001b[0;36m__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1049\u001b[0m         \u001b[0;34m\"\"\"Return the next line (Py2 Compatibility).\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1050\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1051\u001b[0;31m         \u001b[0mread\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbamnostic\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAlignedSegment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1052\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1053\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/bamnostic/core.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, _io)\u001b[0m\n\u001b[1;32m    218\u001b[0m         \"\"\"\n\u001b[1;32m    219\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_io\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_io\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 220\u001b[0;31m         \u001b[0mbsize_buffer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_io\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    221\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m             \u001b[0mblock_size\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0munpack_int32\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbsize_buffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/anaconda3/lib/python3.8/site-packages/bamnostic/bgzf.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    420\u001b[0m             \u001b[0;31m# to multiple blocks, or fully contained within one.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    421\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 422\u001b[0;31m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_within_block_offset\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msize\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_buffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    423\u001b[0m                     \u001b[0;31m# This may leave us right at the end of a block\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    424\u001b[0m                     \u001b[0;31m# (lazy loading, don't load the next block unless we have too)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: object of type 'NoneType' has no len()"
     ]
    }
   ],
   "source": [
    "next(bam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "# lets align to a genome of. size 10 a read of 3 nucleotides starting at pos (0-based) 4:\n",
    "read.seq ACT\n",
    "read.pos 4\n",
    "genome_pos  0 0 0 0 0 0 0 0 0 0    # number of reads that align at. each position\n",
    "\n",
    "____\n",
    "read      :         A C T\n",
    "genome_pos: 0 0 0 0 1 1 1 0 0 0 0 \n",
    "index     : 0 1 2 3 4 5 6 7 8 9 10\n",
    "_____________\n",
    "\n",
    "# lets align to a genome of size 10 another read of 3 nucleotides starting at pos (0-based) 5:\n",
    "read.seq CAC\n",
    "read.pos 5\n",
    "____\n",
    "read      :         A C T\n",
    "read1     :           C A C\n",
    "genome_pos: 0 0 0 0 1 2 2 1 0 0 0\n",
    "index     : 0 1 2 3 4 5 6 7 8 9 10\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "print(genome_pos[:50])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "# now let's do that with some code\n",
    "# aligning 1 read\n",
    "\n",
    "with bs.AlignmentFile(bs.example_bam, 'rb') as bam:\n",
    "    no_reads = 0\n",
    "    for read in bam:\n",
    "        # printing another read so increase the number\n",
    "        no_reads += 1 \n",
    "        # print(read.seq, read.pos)\n",
    "        # we need to increase the values in the list at indices \n",
    "        # starting with read.pos and ending after we went through \n",
    "        # each nucleotide in the read sequence\n",
    "        for position in range(read.pos, read.pos + len(read.seq)):\n",
    "            genome_pos[position] += 1\n",
    "         # when you reached 1 read\n",
    "        if no_reads >= 1:\n",
    "            # exit the loop\n",
    "            break \n",
    "            \n",
    "print(genome_pos[80:170])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset the genome_positions to start over\n",
    "# create empty list of positions\n",
    "genome_pos = [] \n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    # read and ignore the first line because it does not have nucleotides\n",
    "    genome_file.readline()\n",
    "    for line in genome_file:\n",
    "        # go through each nucleotide one at a time, \n",
    "        # strip new line at the end of the line\n",
    "        for nc in line.strip():\n",
    "            # add a 0 for each nucleotide\n",
    "            genome_pos.append(0)\n",
    "        # stop to see what we have on the first line of nucleotides\n",
    "        # break "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "print(genome_pos[:170])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align 10 reads\n",
    "with bs.AlignmentFile(bs.example_bam, 'rb') as bam:\n",
    "    no_reads = 0\n",
    "    for read in bam:\n",
    "        # printing another read so increase the number\n",
    "        no_reads += 1 \n",
    "        # print(read.seq, read.pos)\n",
    "        # we need to increase the values in the list at indices \n",
    "        # starting with read.pos and ending after we went through \n",
    "        # each nucleotide in the read sequence\n",
    "        for position in range(read.pos, read.pos + len(read.seq)):\n",
    "            genome_pos[position] += 1\n",
    "         # when you reached 10 reads\n",
    "        if no_reads >= 10:\n",
    "            # exit the loop\n",
    "            break "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 4, 4, 5, 7, 7, 7, 7, 7, 7, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 8, 8, 8, 7, 7, 7, 7, 7, 7, 7, 6, 6, 5, 3, 3, 3, 3, 3, 3, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "print(genome_pos[80:170])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3, 4, 4, 4, 5, 8, 12, 14, 14, 14, 15, 16, 16, 16, 16, 17, 18, 19, 19, 19, 19, 19, 19, 20, 21, 22, 22, 22, 22, 23, 23, 23, 24, 25, 25, 23, 23, 23, 24, 24, 23, 20, 20, 21, 23, 24, 24, 25, 25, 27, 26, 26, 25, 26, 26, 27, 29, 30, 29, 28, 27, 27, 29, 32, 33, 34, 34, 35, 35, 35, 35, 35, 35, 33, 32, 30, 32, 29, 29, 27, 26, 26, 25, 27, 25, 26, 27, 27, 27, 27, 29, 27, 28, 29, 32, 32, 33, 33, 29, 29, 28, 28, 27, 28, 29, 31, 32, 32, 33, 34, 34, 33, 35, 38, 38, 37, 37, 36, 36, 37, 38, 37, 40, 42, 42, 41, 41, 40, 40, 37, 38, 36, 36, 36, 38, 38, 38, 38, 37, 37, 36, 36, 34, 33, 32, 31, 31, 31, 28, 29, 29, 29, 30, 30, 29, 27, 26, 23, 23, 22, 22, 22, 24, 24, 27, 26, 28, 27, 28, 25, 26, 26, 26, 27, 28, 29, 28, 29, 29, 30, 33, 35, 33, 34, 35, 35, 37, 40, 38, 37, 37, 38, 40, 39, 41, 41, 43, 41, 41, 42, 44, 43, 45, 44, 46, 47, 49, 48, 52, 52, 53, 54, 58, 59, 63, 64, 63, 64, 64, 65, 68, 70, 71, 76, 78, 79, 81, 83, 87, 88, 89, 89, 89, 91, 89, 91, 94, 93, 94, 96, 93, 93, 93, 89, 90, 92, 94, 93, 92, 95, 91, 90, 89, 89, 88, 88, 84, 81, 82, 77, 78, 78, 78, 74, 75, 76, 75, 74, 77, 80, 80, 78, 79, 78, 77, 75, 78, 80, 81, 81, 77, 76, 76, 76, 75, 74, 78, 80, 79, 79, 83, 82, 82, 81, 84, 88, 89, 86, 92, 89]\n"
     ]
    }
   ],
   "source": [
    "# align all reads\n",
    "\n",
    "# reset the genome_positions to start over\n",
    "# create empty list of positions\n",
    "genome_pos = [] \n",
    "with open(\"genome.fa\") as genome_file:\n",
    "    # read and ignore the first line because it does not have nucleotides\n",
    "    genome_file.readline()\n",
    "    for line in genome_file:\n",
    "        # go through each nucleotide one at a time, \n",
    "        # strip new line at the end of the line\n",
    "        for nc in line.strip():\n",
    "            # add a 0 for each nucleotide\n",
    "            genome_pos.append(0)\n",
    "        # stop to see what we have on the first line of nucleotides\n",
    "        # break \n",
    "\n",
    "# count the reads that align using all reads in the file\n",
    "# no need to keep count of the number of reads\n",
    "with bs.AlignmentFile(bs.example_bam, 'rb') as bam:\n",
    "    for read in bam:\n",
    "        # we need to increase the values in the list at indices \n",
    "        # starting with read.pos and ending after we went through \n",
    "        # each nucleotide in the read sequence\n",
    "        for position in range(read.pos, read.pos + len(read.seq)):\n",
    "            genome_pos[position] += 1\n",
    "\n",
    "            \n",
    "# check to see our genome position list\n",
    "# how many reads aligned at each position\n",
    "print(genome_pos[:300])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATCTTTGGAGTAACTATTATTTCGCCCCTTTTGTTTGCTGCATATCGCCCCGCTCTCTGCATACACGATTGGATAATGACCAAAGCAAGGTTTAATACGC -1\n"
     ]
    }
   ],
   "source": [
    "# we used the example file from the package not the sample.bam  provided in the folder with the notebook\n",
    "\n",
    "with bs.AlignmentFile('sample.bam', 'rb') as bam:\n",
    "    for read in bam:\n",
    "        print(read.seq, read.pos)\n",
    "        break # stop after the first read\n",
    "\n",
    "# the read aligns to a negative position -1 => impossible position \n",
    "# unusable file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
